---
title: "Temporal trial"
output: html_document
---


# Bacterial community analysis
### **Becker *et al.* "Spatio-temporal variation"**

-----------------

This document contains all statistical analyses conducted for the manuscript. Note that due to the random iterative nature of some analyses (such as PERMANOVA, CAP & PCA) some of the figure parameters will change slightly during reanalysis, though core results will remain essentially unchanged. 

All data to reproduce analysis can be found here: `https://github.com/mfbeuq/becker_etal_spatiotemporal`

-----------------

Load necessary ecological analysis libraries. 

```{r,cache=TRUE, message=FALSE}
library(phyloseq)
library(microbiome)
library(ggord)
library(metagMisc)
library(ggpubr)
library(FSA)
library(knitr)
library(rmarkdown)
library(ape)
library(vegan)
library(philr)
library(compositions)
library(qiime2R)
library(plyr)
library(dplyr)
library(tidyr)
library(PMCMR)
library(tibble)
library(viridis)
library(gridExtra)
library(AcidPlots)
library(grid)
library(colorRamps)
library(rstatix)
library(dunn.test)
library(pairwiseAdonis)
library(dplyr)
library(ANCOMBC)
library(nlme)
library(tidyverse)
library(compositions)
library(readr)
library(DT)
library(matrixStats)
library(pheatmap)
library(RColorBrewer)
library(dendsort)
library(ComplexHeatmap)
library(circlize)
library(round)

set.seed(225)
```


This beginning workspace contains: 

* `ps_fp`: phyloseq object including the metadata, and the ASV table, the taxonomy table and the phylogenetic tree, all three created by QIIME2
* `psL.meta`: Shannon diversity index data of the loosely associated (L) root microbiota sub data set
* `psT.meta`: Shannon diversity index data of the tightly associated (T) root microbiota sub data set
* `RA.dist.L`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota sub data set
* `RA.dist.T`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota sub data set


```{r,cache=TRUE}
load('temporal_trial_diversity.RData')
```

Load functions necessary:
```{r,cache=TRUE}
source('functions.R')
```


--------------------

### Alpha diversity statistics:

**Loosely associated root microbiome:**  
The mean Shannon diversity for the L-compartment is `r mean(psL.meta$shannon)` with a standard deviation of `r sd(psL.meta$shannon)`

The following calculates the Kruskal-Wallis test statistics for the L-compartment:

```{r, cache=TRUE}
ps.meta.tib <- psL.meta %>%
  select(c("plant","timepoint","season","row","shannon")) %>% 
  as_tibble()
ps.meta.tib <- ps.meta.tib %>%
  pivot_longer(-c(plant,timepoint,season,row), names_to = "variables", values_to = "value")
#run kw test
tmp2 <- ps.meta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ plant) %>%
  add_column(factor = "Tree")
tmp3 <- ps.meta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ timepoint) %>%
  add_column(factor = "Sampling Timepoint")
tmp5 <- ps.meta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ row) %>%
  add_column(factor = "Row")
dplyr::bind_rows(as_tibble(tmp2),as_tibble(tmp3),as_tibble(tmp5))
```

Dunn's test for multiple comparisons was not necessary as no factor was significant:

<p>&nbsp;</p>
**Tightly associated root microbiome:**  
The mean Shannon diversity for the T-compartment is `r mean(psT.meta$shannon)` with a standard deviation of `r sd(psT.meta$shannon)`

The following calculates the Kruskal-Wallis test statistics for the T-compartment:

```{r, cache=TRUE}
ps.meta.tib <- psT.meta %>%
  select(c("plant","timepoint","season","row","shannon")) %>% 
  as_tibble()
ps.meta.tib <- ps.meta.tib %>%
  pivot_longer(-c(plant,timepoint,season,row), names_to = "variables", values_to = "value")
#run kw test
tmp2 <- ps.meta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ plant) %>%
  add_column(factor = "Tree")
tmp3 <- ps.meta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ timepoint) %>%
  add_column(factor = "Sampling Timepoint")
tmp5 <- ps.meta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ row) %>%
  add_column(factor = "Row")
dplyr::bind_rows(as_tibble(tmp2),as_tibble(tmp3),as_tibble(tmp5))
```


The following performs Dunn's test for multiple comparisons between the sampling timepoint:
```{r, cache=TRUE, cache=TRUE, warning = FALSE}
dunn.test(psT.meta$shannon , psT.meta$date, method="bh", alpha=0.1, list=T, table = F)
```
<p>&nbsp;</p>


### Alpha diversity plots:

This creates a Shannon diversity index line plot for both L- and T-compartment over the course of one year including an error bar (**figure 4 A**)

```{r, cache=TRUE}
shannon_df <- shannon$data %>% rownames_to_column("SampleID") # this moves the sample names to a new column that matches the metadata and allows them to be merged
date_labels <- c("23/05/2018", "14/06/2018", "04/07/2018", "16/07/2018", "06/08/2018", "23/08/2018", "18/09/2018", "25/10/2018","19/12/2018", "27/02/2019", "18/03/2019", "17/04/2019")

#merge metadata and remove empty shannon values
tmp <- q_meta %>% 
  left_join(shannon_df) %>%
  filter(!is.na(shannon_entropy)) 
tmp$date_format <- as.character(tmp$date) %>%
  as.Date(format="%d/%m/%Y")
p1 <- tmp %>%
  ggplot(aes(x=`date_format`, y=shannon_entropy, color=`compartment`)) +
  stat_summary(geom="errorbar", fun.data=mean_se, width=0.2, size=1) +
  stat_summary(geom="line", fun.data=mean_se, size=2) +
  stat_summary(geom="point", fun.data=mean_se, size=3) +
  xlab("") +
  ylab("Shannon Diversity") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=20)) +
  scale_x_date(date_labels = "%m-%Y", date_breaks = "1 month",
               limits = as.Date(c("05-2018","05-2019"), format = "%m-%Y"), expand = c(0,25)) +
  scale_color_manual(values = c("black", "red"), name="Compartment") # use different color scale which is color blind friendly
p1 + theme(legend.position = c(0.2,0.15),
           legend.background = element_rect(fill=NA, size=0.5, linetype="solid", colour ="black"),
           legend.key.size = unit(1, "cm"),
           legend.key.width = unit(1,"cm"))
```


--------------------

### Beta Diversity:

**Loosely associated root microbiome:**  
First, the distance matrix calculated by QIIME2 needs to be merged with the metadata file:
```{r, cache=TRUE}
RA.dist <- RA.dist.L #using a temporary variable
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)[,c(2:17)]
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree and Date
```{r, cache=TRUE}
cap <- capscale(formula = RA.dist ~ de$plant + de$date, data = de, comm = RA.dist, sqrt.dist = F)
```

Furthermore, PERMANOVA is calculated on the distance matrix
```{r, cache=TRUE}
adonis2(RA.dist ~ de$plant + de$date + de$row, data = de)
```

In addition, pairwise PERMANOVA is calculated for the sampling timepoint factor
```{r, cache=TRUE, error=FALSE, message=FALSE}
pa.tp.L <- pairwise.adonis(x=RA.dist, factors=de[,"timepoint_cat"], perm = 999, p.adjust.m='fdr')
pa.tp.L
```

The CAP plot for the L-compartment by using the the first two axes of CAP
```{r, cache=TRUE}
x <- as.data.frame(scores(cap, display = "sites", choices = c(1,2,3))) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID

test <- tmp %>% 
  group_by(date) %>% 
  summarise_at(vars(matches("CAP")), mean) %>%
  select(CAP1, CAP2)

p1 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = date)) +
  geom_point(alpha=0.9, size=9) +
  scale_color_manual(name="Date", values= blueYellow(n = 12L)) +
  theme(legend.position="none", font=c("white")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  geom_point(data=tmp %>% 
               group_by(date) %>% 
               summarise_at(vars(matches("CAP")), mean),
             size=9, shape=8, show.legend =F) +
  geom_path(data=test , color ="red") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5)) + theme_bw() + 
  theme(legend.title = element_text(size = 30), 
        title = element_text(face = "bold", size = 24, colour = "black"),
        legend.text = element_text(size = 22), 
        legend.key.size = unit(2, "cm"),
        legend.key.width = unit(1,"cm"), 
        axis.text.x = element_text(face = "bold", size = 20, colour = "black"),
        axis.text.y = element_text(face = "bold", size = 20, colour = "black"), 
        axis.title = element_text(face = "bold", size = 20, colour = "black"),
        strip.text = element_text(size = 20))
print(p1)
```


<p>&nbsp;</p>

**Tightly associated root microbiome:**  
First, the distance matrix calculated by QIIME2 needs to be merged with the metadata file:
```{r, cache=TRUE}
RA.dist <- RA.dist.T #using a temporary variable
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)[,c(2:17)]
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree and Date
```{r, cache=TRUE}
cap <- capscale(formula = RA.dist ~ de$plant + de$date, data = de, comm = RA.dist, sqrt.dist = F)
```

Furthermore, PERMANOVA is calculated on the distance matrix
```{r, cache=TRUE}
adonis2(RA.dist ~ de$plant + de$date + de$row, data = de)
```

In addition, pairwise PERMANOVA is calculated for the sampling timepoint factor
```{r, cache=TRUE, error=FALSE, message=FALSE}
pa.tp.T <- pairwise.adonis(x=RA.dist, factors=de[,"timepoint_cat"], perm = 999, p.adjust.m='fdr')
pa.tp.T
```

In addition, pairwise PERMANOVA is calculated for the tree
```{r, cache=TRUE, tidy=TRUE}
pa.tree.T <- pairwise.adonis(x=RA.dist, factors=de[,"plant"], perm = 999, p.adjust.m='fdr')
pa.tree.T
```

The CAP plot for the L-compartment by using the the first two axes of CAP
```{r, cache=TRUE}
x <- as.data.frame(scores(cap, display = "sites", choices = c(1,2,3))) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID

test <- tmp %>% 
  group_by(date) %>% 
  summarise_at(vars(matches("CAP")), mean) %>%
  select(CAP1, CAP2)

p2 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = date)) +
  geom_point(alpha=0.9, size=9) +
  scale_color_manual(name="Date", values= blueYellow(n = 12L)) +
  theme(legend.position="none", font=c("white")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  geom_point(data=tmp %>% 
               group_by(date) %>% 
               summarise_at(vars(matches("CAP")), mean),
             size=9, shape=8, show.legend =F) +
  geom_path(data=test , color ="red") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5)) + theme_bw() + 
  theme(legend.title = element_text(size = 30), 
        title = element_text(face = "bold", size = 24, colour = "black"),
        legend.text = element_text(size = 22), 
        legend.key.size = unit(2, "cm"),
        legend.key.width = unit(1,"cm"), 
        axis.text.x = element_text(face = "bold", size = 20, colour = "black"),
        axis.text.y = element_text(face = "bold", size = 20, colour = "black"), 
        axis.title = element_text(face = "bold", size = 20, colour = "black"),
        strip.text = element_text(size = 20))
print(p2)
```



### Manuscript plot:

The following part creates the combined plot shown in **figure 4 B** in the manuscript but is not executed here

```{r, eval=FALSE}
legend2 <- cowplot::get_legend(p1)
c1 <- ggpar(p1, ylim=c(-1.3,2.5), xlim=c(-1.2,1.6), legend = "", legend.title = "", font.x = c(20, "bold"), font.y = c(20, "bold"), font.tickslab = c(18, color="black")) +
  annotate(geom="text", x=1, y=2.2, label="(A) L-compartment", color="black", size=8)
c2 <- ggpar(p2, ylim=c(-1.3,2.5), xlim=c(-1.2,1.6), legend = "", legend.title = "", font.x = c(20, "bold"), font.y = c(20, "bold"), font.tickslab = c(18, color="black")) +
  annotate(geom="text", x=1, y=2.2, label="(B) T-compartment", color="black", size=8)
figure <- ggarrange(c1, c2, legend2, widths = c(2.5,2.5,1),
                    ncol = 3, nrow = 1)
png("CAP_DEICODE_combined_sampletime.png", width = 1400, height = 600)
print(figure)
dev.off()
```

<p>&nbsp;</p>

--------------------

### Differential abundance analyis:

Pull workspace from QIIME2 and additional data

This beginning workspace contains: 

* `ps_fp`: phyloseq object including the metadata, and the ASV table, the taxonomy table and the phylogenetic tree, all three created by QIIME2
* `p.colors`: a list that codes the different phyla in different colors in the plot 


```{r,cache=TRUE}
load('temporal_trial_DA.RData')
```

The following creates text files which contain the statistical analyses
```{r, eval=FALSE}
# Set as a loop to do the analysis for both compartments
for (comp in c("L","T")){
  
  #read in the phyloseq object and filter for samples from the corresponding compartment
  pseq <- subset_samples(ps, compartment==comp) 
  #set variable of interest
  genus_data <- pseq 
  #set taxonomic level
  var1 <- "timepoint_cat"
  
  # S1 --------------------------------------------------------------------
  var2 <- "S1"
  #set the order of factors and change them to get every pairwise comparison
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                   c("TP1","TP2","TP3","TP4","TP5","TP6","TP7","TP8","TP9","TP10","TP11","TP12"))
  
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP1 - TP2", "TP1 - TP3", "TP1 - TP4", "TP1 - TP5", "TP1 - TP6", "TP1 - TP7",
               "TP1 - TP8", "TP1 - TP9", "TP1 - TP10", "TP1 - TP11", "TP1 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 

  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # S2 --------------------------------------------------------------------
  var2 <- "S2"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                c("TP2","TP1","TP3","TP4","TP5","TP6","TP7","TP8","TP9","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP2 - TP1", "TP2 - TP3", "TP2 - TP4", "TP2 - TP5", "TP2 - TP6", "TP2 - TP7",
               "TP2 - TP8", "TP2 - TP9", "TP2 - TP10", "TP2 - TP11", "TP2 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # S3 --------------------------------------------------------------------
  var2 <- "S3"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP3","TP1","TP2","TP4","TP5","TP6","TP7","TP8","TP9","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP3 - TP1", "TP3 - TP2", "TP3 - TP4", "TP3 - TP5", "TP3 - TP6", "TP3 - TP7",
               "TP3 - TP8", "TP3 - TP9", "TP3 - TP10", "TP3 - TP11", "TP3 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # S4 --------------------------------------------------------------------
  var2 <- "S4"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP4","TP1","TP2","TP3","TP5","TP6","TP7","TP8","TP9","TP10","TP11","TP12"))

  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP4 - TP1", "TP4 - TP2", "TP4 - TP3", "TP4 - TP5", "TP4 - TP6", "TP4 - TP7",
               "TP4 - TP8", "TP4 - TP9", "TP4 - TP10", "T14 - TP11", "TP4 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # S5 --------------------------------------------------------------------
  var2 <- "S5"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP5","TP1","TP2","TP3","TP4","TP6","TP7","TP8","TP9","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP5 - TP1", "TP5 - TP2", "TP5 - TP3", "TP5 - TP4", "TP5 - TP6", "TP5 - TP7",
               "TP5 - TP8", "TP5 - TP9", "TP5 - TP10", "TP5 - TP11", "TP5 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # S6 --------------------------------------------------------------------
  var2 <- "S6"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP6","TP1","TP2","TP3","TP4","TP5","TP7","TP8","TP9","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP6 - TP1", "TP6 - TP2", "TP6 - TP3", "TP6 - TP4", "TP6 - TP5", "TP6 - TP7",
               "TP6 - TP8", "TP6 - TP9", "TP6 - TP10", "TP6 - TP11", "TP6 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # S7 --------------------------------------------------------------------
  var2 <- "S7"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP7","TP1","TP2","TP3","TP4","TP5","TP6","TP8","TP9","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP7 - TP1", "TP7 - TP2", "TP7 - TP3", "TP7 - TP4", "TP7 - TP5", "TP7 - TP6", 
               "TP7 - TP8", "TP7 - TP9", "TP7 - TP10", "TP7 - TP11", "TP7 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # S8 --------------------------------------------------------------------
  var2 <- "S8"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP8","TP1","TP2","TP3","TP4","TP5","TP6","TP7","TP9","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP8 - TP1", "TP8 - TP2", "TP8 - TP3", "TP8 - TP4", "TP8 - TP5", "TP8 - TP6", "TP8 - TP7",
               "TP8 - TP9", "TP8 - TP10", "TP8 - TP11", "TP8 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # S9 --------------------------------------------------------------------
  var2 <- "S9"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP9","TP1","TP2","TP3","TP4","TP5","TP6","TP7","TP8","TP10","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP9 - TP1", "TP9 - TP2", "TP9 - TP3", "TP9 - TP4", "TP9 - TP5", "TP9 - TP6", "TP9 - TP7",
               "TP9 - TP8", "TP9 - TP10", "TP9 - TP11", "TP9 - TP12")  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # S10 --------------------------------------------------------------------
  var2 <- "S10"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP10","TP1","TP2","TP3","TP4","TP5","TP6","TP7","TP8","TP9","TP11","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP10 - TP1", "TP10 - TP2", "TP10 - TP3", "TP10 - TP4", "TP10 - TP5", "TP10 - TP6", "TP10 - TP7",
               "TP10 - TP8", "TP10 - TP9", "TP10 - TP11", "TP10 - TP12")  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # S11 --------------------------------------------------------------------
  var2 <- "S11"
  sample_data(genus_data)$timepoint_cat <- factor(sample_data(genus_data)$timepoint_cat, levels = 
                                                    c("TP11","TP1","TP2","TP3","TP4","TP5","TP6","TP7","TP8","TP9","TP10","TP12"))
  
  #column names
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP11 - TP1", "TP11 - TP2", "TP11 - TP3", "TP11 - TP4", "TP11 - TP5", "TP11 - TP6",
               "TP11 - TP7", "TP11 - TP8", "TP11 - TP9", "TP11 - TP10", "TP11 - TP12")
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #get Coefficients and change the column names of each sub data frame
  tab_coef = res$W
  colnames(tab_coef) 
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  #only filter the test statistics and the coefficients
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
}

# now combine the two data sets as a loop

for (comp in c("L","T")){
  
  var1="timepoint_cat"; tax="ASV"
  var2="S1";L1 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP1...TP2.x","TP1...TP2.y")]
  var2="S2";L2 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP2...TP3.x","TP2...TP3.y")]
  var2="S3";L3 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP3...TP4.x","TP3...TP4.y")]
  var2="S4";L4 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP4...TP5.x","TP4...TP5.y")]
  var2="S5";L5 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP5...TP6.x","TP5...TP6.y")]
  var2="S6";L6 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP6...TP7.x","TP6...TP7.y")]
  var2="S7";L7 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP7...TP8.x","TP7...TP8.y")]
  var2="S8";L8 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP8...TP9.x","TP8...TP9.y")]
  var2="S9";L9 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP9...TP10.x","TP9...TP10.y")]
  var2="S10";L10 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP10...TP11.x","TP10...TP11.y")]
  var2="S11";L11 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP11...TP12.x","TP11...TP12.y")]
  var2="S1";L12 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c("Row.names","TP1...TP12.x","TP1...TP12.y")]
  
  #merge the individual comparisons into a single data frame
  tmp_M <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(L1,L2,L3,L4,L5,L6,L7,L8,L9,L10,L11,L12))
  rownames(tmp_M) <- tmp_M$Row.names
  tmp_M$Row.names = NULL
  head(tmp_M)
  
  #add the taxonomy and the mean relative abundance with its standard deviation
  pseq <- subset_samples(ps, compartment==comp) 
  PGroup <- transform_sample_counts(pseq, function(x)100* x / sum(x))
  OTUg <- otu_table(PGroup)
  TAXg <- tax_table(PGroup)
  AverageD <- as.data.frame(rowMeans(OTUg))
  names(AverageD) <- c("Mean")
  SD <- as.data.frame(rowSds(OTUg),na.rm = T)
  names(SD) <- c("SD")
  tmp <- cbind(AverageD,SD)
  GTable <- merge(TAXg, tmp, by=0, all=TRUE)
  rownames(GTable) <- GTable$Row.names
  GTable$Row.names = NULL
  GTable <- GTable[,c(9,10)]
  head(GTable)
  
  tmp5 <- merge(tmp_M, GTable, by=0) 
  row.names(tmp5) <- tmp5$Row.names
  tmp5$Row.names = NULL

  #export the results to a .csv file
  write.table(tmp5, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"_results_timepoint.csv"), sep="\t")
}


#Merge L and T into a single data frame
tmp_L <- read.table(file=paste0("../stat_results/ANCOMBC/pairwise/",tax,"/L_results_timepoint.csv"))
colnames(tmp_L) <- paste("L", colnames(tmp_L), sep = " - ")
tmp_T <- read.table(file=paste0("../stat_results/ANCOMBC/pairwise/",tax,"/T_results_timepoint.csv"))
colnames(tmp_T) <- paste("T", colnames(tmp_T), sep = " - ")
tmp <- merge(tmp_L, tmp_T, by=0, all= T) 
rownames(tmp) <- tmp$Row.names
tmp$Row.names <- NULL
head(tmp)

#check for correct taxonomic assignment
pseq <- ps
tmp3 <- as.data.frame(tax_table(pseq))[,-c(1,7,8)]
tmp4 <- merge(tmp3, tmp, by=0) 
rownames(tmp4) <- tmp4$Row.names
tmp4$Row.names = NULL

#filter out for ASVs with a relative abundance >= 0.3 %.
ancom <- tmp4[ which(tmp4$`L - Mean` >= 0.3 | tmp4$`T - Mean` >= 0.3), ]
colnames(ancom)[1:5] <- c("Phylum", "Class", "Order", "Family", "Genus")

#add phyla colors and add name row
ancom <- merge(ancom, p.colors, by="Phylum", all.y = T, )
ancom$Genus[ancom$Genus=="uncultured"] <- "Unclassified"
ancom$Genus[is.na(ancom$Genus)] <- "Unclassified"
ancom <- add_column(ancom, Name = ancom$Genus, .after = "Genus")

#chose only significant rows!
tmp <- ancom
colnames(tmp)
for (i in c(7,9,11,13,15,17,19,21,23,25,27,29,33,35,37,39,41,43,45,47,49,51,53,55)){
  tmp[,i][is.na(tmp[,i])] <- FALSE
}

tmp <- tmp[(rowSums(tmp[,c(7,9,11,13,15,17,19,21,23,25,27,29,33,35,37,39,41,43,45,47,49,51,53,55)])!=0),]
head(tmp)

#filter for L-samples and rename ASV names (some have to be edited manually!)
exp_L <- tmp[(rowSums(tmp[,c(7,9,11,13,15,17,19,21,23,25,27,29)])!=0),]
exp_L <- exp_L[,c(1:32,59)]
exp_L <- rename.ancombc.output(exp_L)

#filter for T-samples and rename ASV names (some have to be edited manually!)
exp_T <- tmp[(rowSums(tmp[,c(33,35,37,39,41,43,45,47,49,51,53,55)])!=0),]
exp_T <- exp_T[,c(1:6,33:59)]
exp_T <- rename.ancombc.output(exp_T)


###### Do the same analysis but with the trees as the main factor in order to compare them:
for (comp in c("L","T")){
  
  pseq <- subset_samples(ps, compartment==comp) 
  genus_data <- pseq
  var1 <- "plant"
  
  # T1 --------------------------------------------------------------------
  var2 <- "T1"
  sample_data(genus_data)$plant <- factor(sample_data(genus_data)$plant, levels = 
                                                c("B1","B2","B3","B4","B5","B6"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("B1 - B2","B1 - B3","B1 - B4","B1 - B5","B1 - B6",
               "TP1 - TP2", "TP1 - TP3", "TP1 - TP4", "TP1 - TP5", "TP1 - TP6", "TP1 - TP7",
               "TP1 - TP8", "TP1 - TP9", "TP1 - TP10", "TP1 - TP11", "TP1 - TP12")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # T2 --------------------------------------------------------------------
  var2 <- "T2"
  sample_data(genus_data)$plant <- factor(sample_data(genus_data)$plant, levels = 
                                            c("B2","B1","B3","B4","B5","B6"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("B2 - B1","B2 - B3","B2 - B4","B2 - B5","B2 - B6",
               "TP1 - TP2", "TP1 - TP3", "TP1 - TP4", "TP1 - TP5", "TP1 - TP6", "TP1 - TP7",
               "TP1 - TP8", "TP1 - TP9", "TP1 - TP10", "TP1 - TP11", "TP1 - TP12")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # T3 --------------------------------------------------------------------
  var2 <- "T3"
  sample_data(genus_data)$plant <- factor(sample_data(genus_data)$plant, levels = 
                                            c("B3","B1","B2","B4","B5","B6"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("B3 - B1","B3 - B2","B3 - B4","B3 - B5","B3 - B6",
               "TP1 - TP2", "TP1 - TP3", "TP1 - TP4", "TP1 - TP5", "TP1 - TP6", "TP1 - TP7",
               "TP1 - TP8", "TP1 - TP9", "TP1 - TP10", "TP1 - TP11", "TP1 - TP12")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # T4 --------------------------------------------------------------------
  var2 <- "T4"
  sample_data(genus_data)$plant <- factor(sample_data(genus_data)$plant, levels = 
                                            c("B4","B1","B2","B3","B5","B6"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("B4 - B1","B4 - B2","B4 - B3","B4 - B5","B4 - B6",
               "TP1 - TP2", "TP1 - TP3", "TP1 - TP4", "TP1 - TP5", "TP1 - TP6", "TP1 - TP7",
               "TP1 - TP8", "TP1 - TP9", "TP1 - TP10", "TP1 - TP11", "TP1 - TP12")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # T5 --------------------------------------------------------------------
  var2 <- "T5"
  sample_data(genus_data)$plant <- factor(sample_data(genus_data)$plant, levels = 
                                            c("B5","B1","B2","B3","B4","B6"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "plant + timepoint_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("B5 - B1","B5 - B2","B5 - B3","B5 - B4","B5 - B6",
               "TP1 - TP2", "TP1 - TP3", "TP1 - TP4", "TP1 - TP5", "TP1 - TP6", "TP1 - TP7",
               "TP1 - TP8", "TP1 - TP9", "TP1 - TP10", "TP1 - TP11", "TP1 - TP12")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0) 
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  }

# combine the data sets --------------------------------------------------------

for (comp in c("L","T")){
  
  pseq <- subset_samples(ps, compartment==comp) 
  genus_data <- pseq
  var1 <- "plant"; tax="ASV"
  
  var2="T1";T1 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c(1:6,18:22)]
  var2="T2";T2 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c(1:6,18:22)]
  var2="T3";T3 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c(1:6,18:22)]
  var2="T4";T4 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c(1:6,18:22)]
  var2="T5";T5 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))[,c(1:6,18:22)]
  
  tmp <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(T1,T2,T3,T4,T5))
  rownames(tmp) <- tmp$Row.names
  tmp_M <- tmp[,c("B1...B2.x","B1...B3.x","B1...B4.x","B1...B5.x","B1...B6.x",
                "B2...B3.x","B2...B4.x","B2...B5.x","B2...B6.x",
                "B3...B4.x","B3...B5.x","B3...B6.x",
                "B4...B5.x","B4...B6.x",
                "B5...B6.x",
                "B1...B2.y","B1...B3.y","B1...B4.y","B1...B5.y","B1...B6.y",
                "B2...B3.y","B2...B4.y","B2...B5.y","B2...B6.y",
                "B3...B4.y","B3...B5.y","B3...B6.y",
                "B4...B5.y","B4...B6.y",
                "B5...B6.y")]
  head(tmp_M)
  
  pseq <- subset_samples(ps, compartment==comp) 
  PGroup <- transform_sample_counts(pseq, function(x)100* x / sum(x))
  OTUg <- otu_table(PGroup)
  TAXg <- tax_table(PGroup)
  AverageD <- as.data.frame(rowMeans(OTUg))
  names(AverageD) <- c("Mean")
  SD <- as.data.frame(rowSds(OTUg),na.rm = T)
  names(SD) <- c("SD")
  tmp <- cbind(AverageD,SD)
  GTable <- merge(TAXg, tmp, by=0, all=TRUE)
  rownames(GTable) <- GTable$Row.names
  GTable$Row.names = NULL
  GTable <- GTable[,c(9,10)]
  head(GTable)
  
  tmp5 <- merge(tmp_M, GTable, by=0) 
  row.names(tmp5) <- tmp5$Row.names
  tmp5$Row.names = NULL
  head(tmp5)
  write.table(tmp5, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"_results_tree.csv"), sep="\t")
}

#Merge L and T
tmp_L <- read.table(file=paste0("../stat_results/ANCOMBC/pairwise/",tax,"/L_results_tree.csv"))
colnames(tmp_L) <- paste("L", colnames(tmp_L), sep = " - ")
tmp_T <- read.table(file=paste0("../stat_results/ANCOMBC/pairwise/",tax,"/T_results_tree.csv"))
colnames(tmp_T) <- paste("T", colnames(tmp_T), sep = " - ")
tmp <- merge(tmp_L, tmp_T, by=0, all= T) 
rownames(tmp) <- tmp$Row.names
tmp$Row.names <- NULL
head(tmp)

#add taxonomy
pseq <- ps
tmp3 <- as.data.frame(tax_table(pseq))[,-c(1,7,8)]
tmp4 <- merge(tmp3, tmp, by=0) 
rownames(tmp4) <- tmp4$Row.names
tmp4$Row.names = NULL

write_csv(tmp4, file = "../stat_results/ANCOMBC/pairwise/ASV/L_and_T_tree.csv")

ancom <- tmp4[ which(tmp4$`L - Mean` >= 0.3 | tmp4$`T - Mean` >= 0.3), ]
colnames(ancom)[1:5] <- c("Phylum", "Class", "Order", "Family", "Genus")

ancom <- merge(ancom, p.colors, by="Phylum", all.y = T, )
ancom$Genus[ancom$Genus=="uncultured"] <- "Unclassified"
ancom$Genus[is.na(ancom$Genus)] <- "Unclassified"
ancom <- add_column(ancom, Name = ancom$Genus, .after = "Genus")

#chose only significant rows!
tmp <- ancom
colnames(tmp)
for (i in c(7:21,39:53)){
  tmp[,i][is.na(tmp[,i])] <- FALSE
}

tmp <- tmp[(rowSums(tmp[,c(7:21,39:53)])!=0),]
head(tmp)

tax.clean <- tmp
colnames(tax.clean)
tax.clean <- rename.ancombc.output(tax.clean)
```


### Differential abundance plot:

The latter part did the actual differential abundance analysis. Now we have to format the final table. It might be necessary to manually rename some of the ASVs, e.g. when there are two ASVs with the same name. Then, adding a a numbering is necessary. This will create the **figures 5 and S11** 

```{r, eval=FALSE}
# Timepoint --------------------------------
for (comp in c("L","T")){
  if(comp == "L") {
    tmp_L <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_timepoint_significant.csv"), row.names = "Name")
    tmp_L <- tmp_L[,c(1:27,54)]  
    tmp_L <- tmp_L[(rowSums(tmp_L[,c(2,4,6,8,10,12,14,16,18,20,22,24)])!=0),]
    tmp_L <- tmp_L[tmp_L$L...Mean>0.3,]
    write.csv(tmp_L, file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_timepoint_significant_0.3.csv"),  row.names = TRUE)
  } else if(comp == "T") {
    tmp_T <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_timepoint_significant.csv"), row.names = "Name")
    tmp_T <- tmp_T[,c(1,28:54)]  
    tmp_T <- tmp_T[(rowSums(tmp_T[,c(2,4,6,8,10,12,14,16,18,20,22,24)])!=0),]
    tmp_T <- tmp_T[tmp_T$T...Mean>0.3,]
    write.csv(tmp_T, file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/T_timepoint_significant_0.3.csv"),  row.names = TRUE)
  }
}
  
#OR with the other ancombc files with RA < 0.1%
tmp_L <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_timepoint_significant.csv"), row.names = "Name")
tmp_T <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/T_timepoint_significant.csv"), row.names = "Name")

#OR with the other ancombc files with RA < 0.3%
tmp_L <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_timepoint_significant_0.3.csv"), row.names = "X")
tmp_T <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/T_timepoint_significant_0.3.csv"), row.names = "X")
#redo in excel! -> add 

for (comp in c("L","T")){
    
  if(comp == "L") {
    mat.df <- tmp_L
  } else if (comp == "T") {
    mat.df <- tmp_T
  }
  
  str(mat.df)
  colnames(mat.df) <- c("Phylum",
    "q-T1:T2", "TP1 -> TP2", 
    "q-T2:T3", "TP2 -> TP3", 
    "q-T3:T4", "TP3 -> TP4", 
    "q-T4:T5", "TP4 -> TP5", 
    "q-T5:T6", "TP5 -> TP6", 
    "q-T6:T7", "TP6 -> TP7", 
    "q-T7:T8", "TP7 -> TP8", 
    "q-T8:T9", "TP8 -> TP9", 
    "q-T9:T10", "TP9 -> TP10",
    "q-T10:T11", "TP10 -> TP11", 
    "q-T11:T12", "TP11 -> TP12", 
    "q-T1:T12", "TP1 -> TP12", 
    "Mean","SD","P-color")
  
  #IMPORTANT: specify ONLY the columns with the differentials
  col.order <- c("TP1 -> TP2", "TP2 -> TP3", "TP3 -> TP4", "TP4 -> TP5", "TP5 -> TP6", "TP6 -> TP7", 
                 "TP7 -> TP8", "TP8 -> TP9", "TP9 -> TP10", "TP10 -> TP11", "TP11 -> TP12", "TP1 -> TP12")
  mat.diff <- as.matrix(mat.df[,c(3,5,7,9,11,13,15,17,19,21,23,25)])
  sig_mat <- as.matrix(mat.df[,c(2,4,6,8,10,12,14,16,18,20,22,24)])
  #rownames(mat.diff) <- mat.df$Taxonomy
  min_lc <- as.numeric(min(mat.diff, na.rm = T))
  max_lc <- as.numeric(max(mat.diff, na.rm = T))
  colors <- colorRamp2(c(min_lc, 0, max_lc), c("blue", "white", "red"))
  
  hb = rowAnnotation('Mean Abundance [%]' = anno_barplot(as.vector(mat.df$Mean), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1,
                                                         axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 26)), 
                                                         width = unit(10, "cm")), annotation_name_gp = gpar(fontsize = 26, fontface = "bold"),
                     annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  lgd = Legend(at = c(round(min_lc+min_lc*0.05,digits = 2), 0, round(max_lc+max_lc*0.05, digits = 2)),  col_fun = colors, 
               title = "W-value", title_gp = gpar(fontsize = 26, fontface = "bold"), title_gap = unit(0.5, "cm"),
               labels_gp = gpar(col = "black", fontsize = 26, fontface = "bold"), title_position = "topcenter", 
               grid_height = unit(3, "cm"), legend_width = unit(10, "cm"), direction = "horizontal") 
  
  ht = Heatmap(mat.diff , name = "logfold change",
               column_gap = unit(5, "mm"),
               row_split = mat.df$Phylum,
               row_gap = unit(5, "mm"), 
               row_names_gp = gpar(fontsize = 32, fontface = "bold", 
                                   col = mat.df$`P-color`),
               na_col = "grey",
               col = colors,
               rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
               border = TRUE, 
               cluster_rows = FALSE, #remove cluster         
               show_column_dend = FALSE, #remove cluster 
               show_heatmap_legend = FALSE, #r emove legend
               row_names_side = "right",
               row_names_max_width = unit(2, "cm"),
               row_names_rot = 0,
               row_names_centered = FALSE,
               row_title_gp = gpar(fontsize = 28),
               row_title_rot = 0,
               column_title = paste0("ANCOM-BC W-value"), 
               column_title_gp = gpar(fill = "black", col = "white", fontsize = 28, fontface = "bold", border = "black"), 
               column_title_side = "top",
               column_names_max_height = unit(6, "cm"),
               column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
               column_names_rot = 90,
               column_names_centered = TRUE,
               column_order = col.order,
               show_parent_dend_line = FALSE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 if(sig_mat[i, j] == "TRUE")
                   grid.text("*", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               }, 
               right_annotation = c(hb))
  png(paste0("Plots/ANCOM-BC/",comp,"-",tax,"_heatmap_differentials_0.3.png"), width = 2000, height = 300+45*(nrow(mat.diff)))
  draw(ht, padding = unit(c(5, 1, 1, 26), "cm")) # add space for titles
  draw(lgd, x = unit(45, "cm"), y = unit(6, "cm"))
  dev.off()
#
}


# Tree plot ---------------------------------------------------------------

for (comp in c("L","T")){
  if(comp == "L") {
    tmp_L <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_tree_significant.csv"), row.names = "Name")
    tmp_L <- tmp_L[,c(1:33,66)]  
    tmp_L <- tmp_L[(rowSums(tmp_L[,c(2:16)])!=0),]
    tmp_L <- tmp_L[tmp_L$L...Mean>0.3,]
  } else if(comp == "T") {
    tmp_T <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_tree_significant.csv"), row.names = "Name")
    tmp_T <- tmp_T[,c(1,34:66)] 
    tmp_T <- tmp_T[(rowSums(tmp_T[,c(2:16)])!=0),]
    tmp_T <- tmp_T[tmp_T$T...Mean>0.3,]
  }
}

#redo in excel! -> add 

for (comp in c("L","T")){
  
  if(comp == "L") {
    mat.df <- tmp_L
  } else if (comp == "T") {
    mat.df <- tmp_T
  }
  
  colnames(mat.df) <- c("Phylum", 
                        "q-T1:T2","q-T1:T3","q-T1:T4","q-T1:T5","q-T1:T6",
                        "q-T2:T3","q-T2:T4","q-T2:T5","q-T2:T6",
                        "q-T3:T4","q-T3:T5","q-T3:T6",
                        "q-T4:T5","q-T4:T6","q-T5:T6",
                        "T1 -> T2","T1 -> T3","T1 -> T4","T1 -> T5","T1 -> T6",
                        "T2 -> T3","T2 -> T4","T2 -> T5","T2 -> T6",
                        "T3 -> T4","T3 -> T5","T3 -> T6",
                        "T4 -> T5","T4 -> T6","T5 -> T6",
                        "Mean", "SD", "P-color")
  
  #IMPORTANT: specify ONLY the columns with the differentials
  col.order <- c("T1 -> T2","T1 -> T3","T1 -> T4","T1 -> T5","T1 -> T6",
                 "T2 -> T3","T2 -> T4","T2 -> T5","T2 -> T6",
                 "T3 -> T4","T3 -> T5","T3 -> T6",
                 "T4 -> T5","T4 -> T6","T5 -> T6")
  mat.diff <- as.matrix(mat.df[,c(17:31)])
  sig_mat <- as.matrix(mat.df[,c(2:16)])
  #rownames(mat.diff) <- mat.df$Taxonomy
  min_lc <- min(mat.diff, na.rm = T)
  max_lc <- max(mat.diff, na.rm = T)
  colors <- colorRamp2(c(min_lc, 0, max_lc), c("blue", "white", "red"))
  
  hb = rowAnnotation('Mean Abundance [%]' = anno_barplot(as.vector(mat.df$Mean), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1,
                                                         axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 26)), 
                                                         width = unit(10, "cm")), annotation_name_gp = gpar(fontsize = 26, fontface = "bold"),
                     annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  lgd = Legend(at = c(round(min_lc+min_lc*0.05,digits = 2), 0, round(max_lc+max_lc*0.05, digits = 2)),  col_fun = colors, 
               title = "W-value", title_gp = gpar(fontsize = 26, fontface = "bold"), title_gap = unit(0.5, "cm"),
               labels_gp = gpar(col = "black", fontsize = 26, fontface = "bold"), title_position = "topcenter", 
               grid_height = unit(3, "cm"), legend_width = unit(10, "cm"), direction = "horizontal") 
  
  ht = Heatmap(mat.diff , name = "logfold change",
               column_gap = unit(5, "mm"),
               row_split = mat.df$Phylum,
               row_gap = unit(5, "mm"), 
               row_names_gp = gpar(fontsize = 32, fontface = "bold", 
                                   col = mat.df$`P-color`),
               na_col = "grey",
               col = colors,
               rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
               border = TRUE, 
               cluster_rows = FALSE, #remove cluster         
               show_column_dend = FALSE, #remove cluster 
               show_heatmap_legend = FALSE, #r emove legend
               row_names_side = "right",
               row_names_max_width = unit(2, "cm"),
               row_names_rot = 0,
               row_names_centered = FALSE,
               row_title_gp = gpar(fontsize = 28),
               row_title_rot = 0,
               column_title = paste0("ANCOM-BC W-value"), 
               column_title_gp = gpar(fill = "black", col = "white", fontsize = 28, fontface = "bold", border = "black"), 
               column_title_side = "top",
               column_names_max_height = unit(6, "cm"),
               column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
               column_names_rot = 90,
               column_names_centered = TRUE,
               column_order = col.order,
               show_parent_dend_line = FALSE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 if(sig_mat[i, j] == "TRUE")
                   grid.text("*", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               }, 
               right_annotation = c(hb))
  png(paste0("Plots/ANCOM-BC/",comp,"-",tax,"_heatmap_differentials_tree_all.png"), width = 2000, height = 300+45*(nrow(mat.diff)))
  draw(ht, padding = unit(c(5, 1, 1, 26), "cm")) # add space for titles
  draw(lgd, x = unit(45, "cm"), y = unit(6, "cm"))
  dev.off()
}
```



### Heatmaps:
The next part creates the heatmaps **S9** and **S10**. The results from the pairwise PERMANOVA were formatted in excel to transform the data to the right format. The following data data is being imported:

* `pa.tp.L.sig`: the non-adjusted p-values from the pairwise PERMANOVA of the timepoint factor in the L-compartment
* `pa.tp.L.R2`: the R2 values from the pairwise PERMANOVA of the timepoint factor in the L-compartment
* `pa.tp.T.sig`: the non-adjusted p-values from the pairwise PERMANOVA of the timepoint factor in the T-compartment
* `pa.tp.T.R2`: the R2 values from the pairwise PERMANOVA of the timepoint factor in the T-compartment
* `pa.tree.T.sig`: the non-adjusted p-values from the pairwise PERMANOVA of the tree factor in the T-compartment
* `pa.tree.T.R2`: the R2 values from the pairwise PERMANOVA of the tree factor in the T-compartment

```{r,cache=TRUE}
load('temporal_trial_pairwise_permanova_results.RData')
```


```{r, eval=FALSE}
### Plot S10
sig <- pa.tree.T.sig
R2 <- pa.tree.T.R2

colors <- colorRamp2(c(0,0.5, 1), c("beige","darkorange", "darkred"), space = "LAB")

lgd = Legend(at = c(0, 1),  col_fun = colors, 
             title = expression(paste("R"^"2", " value")), title_gp = gpar(fontsize = 26, fontface = "bold"), title_gap = unit(0.5, "cm"),
             labels_gp = gpar(col = "black", fontsize = 26, fontface = "bold"), title_position = "topcenter", 
             grid_height = unit(3, "cm"), legend_width = unit(8, "cm"), direction = "horizontal") 

ht = Heatmap(R2, name = "logfold change",
             column_gap = unit(5, "mm"),
             row_gap = unit(5, "mm"), 
             row_names_gp = gpar(fontsize = 32, fontface = "bold"),
             na_col = "white",
             col = colors,
             rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
             border = TRUE, 
             cluster_rows = FALSE, #remove cluster
             cluster_columns = FALSE,
             show_column_dend = FALSE, #remove cluster 
             show_heatmap_legend = FALSE, #remove legend
             row_names_side = "right",
             row_names_max_width = unit(2, "cm"),
             row_names_rot = 0,
             row_names_centered = FALSE,
             row_title_gp = gpar(fontsize = 28),
             row_title_rot = 0,
             column_title = expression(paste("PERMANOVA R"^"2", " value")),
             column_title_gp = gpar(fill = "black", col = "white", fontsize = 28, fontface = "bold", border = "black"), 
             column_title_side = "top",
             column_names_max_height = unit(6, "cm"),
             column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
             column_names_rot = 90,
             column_names_centered = TRUE,
             show_parent_dend_line = FALSE,
             cell_fun = function(j, i, x, y, width, height, fill) {
               if(sig[i, j] == "**")
                 grid.text("**", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == "*")
                 grid.text("*", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == ".")
                 grid.text(".", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == "0")
                 grid.text("", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
             }
             )
{
  png("../Plots/T_PERMANOVA_trees_heatmap.png", width = 800, height = 800)
  draw(ht, padding = unit(c(1, 1, 1, 1), "cm")) # add space for titles
  draw(lgd, x = unit(20, "cm"), y = unit(22.5, "cm"))
  dev.off()
}

### Plot S9
sig <- pa.tp.L.sig
R2 <- pa.tp.L.R2

colors <- colorRamp2(c(0,0.5, 1), c("beige","darkorange", "darkred"), space = "LAB")

lgd = Legend(at = c(0, 1),  col_fun = colors, 
             title = expression(paste("R"^"2", " value")), title_gp = gpar(fontsize = 26, fontface = "bold"), title_gap = unit(0.5, "cm"),
             labels_gp = gpar(col = "black", fontsize = 26, fontface = "bold"), title_position = "topcenter", 
             grid_height = unit(3, "cm"), legend_width = unit(8, "cm"), direction = "horizontal") 

ht = Heatmap(R2, name = "logfold change",
             column_gap = unit(5, "mm"),
             row_gap = unit(5, "mm"), 
             row_names_gp = gpar(fontsize = 32, fontface = "bold"),
             na_col = "white",
             col = colors,
             rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
             border = TRUE, 
             cluster_rows = FALSE, #remove cluster
             cluster_columns = FALSE,
             show_column_dend = FALSE, #remove cluster 
             show_heatmap_legend = FALSE, #remove legend
             row_names_side = "right",
             row_names_max_width = unit(2, "cm"),
             row_names_rot = 0,
             row_names_centered = FALSE,
             row_title_gp = gpar(fontsize = 28),
             row_title_rot = 0,
             column_title = expression(paste("PERMANOVA R"^"2", " value")),
             column_title_gp = gpar(fill = "black", col = "white", fontsize = 28, fontface = "bold", border = "black"), 
             column_title_side = "top",
             column_names_max_height = unit(6, "cm"),
             column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
             column_names_rot = 90,
             column_names_centered = TRUE,
             show_parent_dend_line = FALSE,
             cell_fun = function(j, i, x, y, width, height, fill) {
               if(sig[i, j] == "**")
                 grid.text("**", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == "*")
                 grid.text("*", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == ".")
                 grid.text(".", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == "0")
                 grid.text("", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
             }
)
{
  png("../Plots/L_PERMANOVA_date_heatmap.png", width = 800, height = 800)
  draw(ht, padding = unit(c(1, 1, 1, 4), "cm")) # add space for titles
  draw(lgd, x = unit(16, "cm"), y = unit(22.5, "cm"))
  dev.off()
}


### T: date ---------------------------------------------------------
sig <- pa.tp.T.sig
R2 <- pa.tp.T.R2

colors <- colorRamp2(c(0,0.5, 1), c("beige","darkorange", "darkred"), space = "LAB")

lgd = Legend(at = c(0, 1),  col_fun = colors, 
             title = expression(paste("R"^"2", " value")), title_gp = gpar(fontsize = 26, fontface = "bold"), title_gap = unit(0.5, "cm"),
             labels_gp = gpar(col = "black", fontsize = 26, fontface = "bold"), title_position = "topcenter", 
             grid_height = unit(3, "cm"), legend_width = unit(8, "cm"), direction = "horizontal") 

ht = Heatmap(R2, name = "logfold change",
             column_gap = unit(5, "mm"),
             row_gap = unit(5, "mm"), 
             row_names_gp = gpar(fontsize = 32, fontface = "bold"),
             na_col = "white",
             col = colors,
             rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
             border = TRUE, 
             cluster_rows = FALSE, #remove cluster
             cluster_columns = FALSE,
             show_column_dend = FALSE, #remove cluster 
             show_heatmap_legend = FALSE, #remove legend
             row_names_side = "right",
             row_names_max_width = unit(2, "cm"),
             row_names_rot = 0,
             row_names_centered = FALSE,
             row_title_gp = gpar(fontsize = 28),
             row_title_rot = 0,
             column_title = expression(paste("PERMANOVA R"^"2", " value")),
             column_title_gp = gpar(fill = "black", col = "white", fontsize = 28, fontface = "bold", border = "black"), 
             column_title_side = "top",
             column_names_max_height = unit(6, "cm"),
             column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
             column_names_rot = 90,
             column_names_centered = TRUE,
             show_parent_dend_line = FALSE,
             cell_fun = function(j, i, x, y, width, height, fill) {
               if(sig[i, j] == "**")
                 grid.text("**", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == "*")
                 grid.text("*", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == ".")
                 grid.text(".", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
               if(sig[i, j] == "0")
                 grid.text("", x, y, gp = gpar(fontsize = 28, fontface = "bold"))
             }
)
{
  png("../Plots/T_PERMANOVA_date_heatmap.png", width = 800, height = 800)
  draw(ht, padding = unit(c(1, 1, 1, 4), "cm")) # add space for titles
  draw(lgd, x = unit(16, "cm"), y = unit(22.5, "cm"))
  dev.off()
}
```
