---
title: "Spatio-temporal trial"
output: html_document
---


# Bacterial community analysis
### **Becker *et al.* "Spatio-temporal variation"**

-----------------

This document contains all statistical analyses conducted for the manuscript. Note that due to the random iterative nature of some analyses (such as PERMANOVA, CAP & PCA) some of the figure parameters will change slightly during reanalysis, though core results will remain essentially unchanged. 

All data to reproduce analysis can be found here: `https://github.com/mfbeuq/becker_etal_spatiotemporal`

-----------------

Load necessary ecological analysis libraries. 

```{r,cache=TRUE, message=FALSE}
library(phyloseq)
library(microbiome)
library(ggord)
library(metagMisc)
library(ggpubr)
library(FSA)
library(knitr)
library(rmarkdown)
library(ape)
library(vegan)
library(philr)
library(compositions)
library(qiime2R)
library(plyr)
library(dplyr)
library(tidyr)
library(PMCMR)
library(tibble)
library(viridis)
library(gridExtra)
library(AcidPlots)
library(grid)
library(colorRamps)
library(rstatix)
library(dunn.test)
library(pairwiseAdonis)
library(dplyr)
library(ANCOMBC)
library(nlme)
library(tidyverse)
library(compositions)
library(readr)
library(DT)
library(matrixStats)
library(pheatmap)
library(RColorBrewer)
library(dendsort)
library(ComplexHeatmap)
library(circlize)
library(round)
library(lme4)
set.seed(225)
```


This beginning workspace contains: 

* `ps_fp`: phyloseq object including the metadata, and the ASV table, the taxonomy table and the phylogenetic tree, all three created by QIIME2
* `q_meta`: metadata file
* `psL.meta`: Shannon diversity index data of the loosely associated (L) root microbiota sub data set
* `psT.meta`: Shannon diversity index data of the tightly associated (T) root microbiota sub data set
* `psLF.meta`: Shannon diversity index data of the loosely associated (T) root microbiota in the fine roots sub data set
* `psLT.meta`: Shannon diversity index data of the loosely associated (T) root microbiota in the thick roots sub data set
* `psTF.meta`: Shannon diversity index data of the tightly associated (T) root microbiota in the fine roots sub data set
* `psTT.meta`: Shannon diversity index data of the tightly associated (T) root microbiota in the thick roots sub data set
* `RA.dist.L`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota sub data set
* `RA.dist.T`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota sub data set
* `RA.dist.L_F`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota in the fine roots sub data set
* `RA.dist.L_T`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota in the thick roots sub data set
* `RA.dist.L_F`: DEICODE distance matrix generated by QIIME2 of the tightly associated (L) root microbiota in the fine roots sub data set
* `RA.dist.L_T`: DEICODE distance matrix generated by QIIME2 of the tightly associated (L) root microbiota in the thick roots sub data set


```{r,cache=TRUE}
load('spatio-temporal_trial_diversity.RData')
```

Load functions necessary:
```{r,cache=TRUE}
source('functions.R')
```


--------------------

### Alpha diversity statistics:

**Loosely associated root microbiome:**  

First the number of days after the first sampling timepoint are added with the first one as zero
```{r, cache=TRUE, warning = FALSE}
psL.meta$days <- psL.meta$sample_time_cat
psL.meta$days[psL.meta$days == "S_4"] <- 0
psL.meta$days[psL.meta$days == "S_5"] <- 25
psL.meta$days[psL.meta$days == "S_6"] <- 76
psL.meta$days[psL.meta$days == "S_7"] <- 152
psL.meta$days <- as.numeric(psL.meta$days)
```

The following fits a linear mixed-effects model on the Shannon diversity as response variable with the factors root section and sampling timepoint as fixed effects and the tree individual as random effect. Afterwards, ANOVA is performed on the model

```{r, cache=TRUE, warning = FALSE}
Model <- lmer(Shannon ~ rootpart * days + (1|tree_cat), data = psL.meta, REML = TRUE)
anova(Model)
```

No significance was detected and thus, no Post-hoc tests were performed.
<p>&nbsp;</p>



**Tightly associated root microbiome:**  

First the number of days after the first sampling timepoint are added with the first one as zero
```{r, cache=TRUE, warning = FALSE}
psT.meta$days <- psT.meta$sample_time_cat
psT.meta$days[psT.meta$days == "S_4"] <- 0
psT.meta$days[psT.meta$days == "S_5"] <- 25
psT.meta$days[psT.meta$days == "S_6"] <- 76
psT.meta$days[psT.meta$days == "S_7"] <- 152
psT.meta$days
```

The following fits a linear mixed-effects model on the Shannon diversity as response variable with the factors root section and sampling timepoint as fixed effects and the tree individual as random effect. Afterwards, ANOVA is performed on the model

```{r, cache=TRUE, warning = FALSE}
Model <- lmer(Shannon ~ rootpart * days + (1|tree_cat), data = psT.meta, REML = TRUE)
anova(Model)
```

No significance was detected and thus, no Post-hoc tests were performed.

<p>&nbsp;</p>


### Beta Diversity:

**Loosely associated root microbiome:**  
First, the distance matrix calculated by QIIME2 needs to be merged with the metadata file:
```{r, cache=TRUE}
RA.dist <- RA.dist.L
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree and Date
```{r, cache=TRUE}
cap <- capscale(formula = RA.dist ~ de$sample_time_cat + de$rootpart + de$tree_cat + cluster1_cat, data = de, comm = RA.dist)
```

Furthermore, PERMANOVA is calculated on the distance matrix
```{r, cache=TRUE}
adonis(RA.dist ~ de$sample_time_cat * de$rootpart * de$cluster1_cat + de$tree_cat, permutations = 999, strata = de$tree_cat)$aov.tab
```

In addition, pairwise PERMANOVA is calculated
```{r, cache=TRUE, error=FALSE, message=FALSE}
pairwise.adonis(x=RA.dist, factors=de[,c(14)], perm = 999) 
```

The CAP plot for the L-compartment by using the the first two axes of CAP
```{r, cache=TRUE}
x <- as.data.frame(scores(cap, display = "sites")) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID
p1 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = sample_time_cat, shape = rootpart)) +
  geom_point(alpha=0.9, size=10) +
  scale_color_manual(name="Sample Time", values= blueYellow(n = 4L), 
                     labels = c("21.03.2019", "15.04.2019", "05.06.2019" ,"20.08.2019")) +
  scale_shape_manual(values=c(15,16), name="Root section", 
                     labels = c("Fine roots","Thick roots")) +
  theme(legend.position="none", font=c("white")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5)) + deicode_theme + theme(legend.key.size = unit(0.8, "cm"))
p1
```


<p>&nbsp;</p>

**Tightly associated root microbiome:**  
First, the distance matrix calculated by QIIME2 needs to be merged with the metadata file:
```{r, cache=TRUE}
RA.dist <- RA.dist.T
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree and Date
```{r, cache=TRUE}
cap <- capscale(formula = RA.dist ~ de$sample_time_cat + de$rootpart + de$tree_cat + cluster1_cat, data = de, comm = RA.dist)
```

Furthermore, PERMANOVA is calculated on the distance matrix
```{r, cache=TRUE}
adonis(RA.dist ~ de$sample_time_cat * de$rootpart * de$cluster1_cat + de$tree_cat, permutations = 999, strata = de$tree_cat)$aov.tab
```

In addition, pairwise PERMANOVA is calculated
```{r, cache=TRUE, error=FALSE, message=FALSE}
pairwise.adonis(x=RA.dist, factors=de[,c(14)], perm = 999)
```


The CAP plot for the L-compartment by using the the first two axes of CAP
```{r, cache=TRUE}
x <- as.data.frame(scores(cap, display = "sites")) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID
p2 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = sample_time_cat, shape = rootpart)) +
  geom_point(alpha=0.9, size=10) +
  scale_color_manual(name="Sample Time", values= blueYellow(n = 4L), 
                     labels = c("21.03.2019", "15.04.2019", "05.06.2019" ,"20.08.2019")) +
  scale_shape_manual(values=c(15,16), name="Root section", 
                     labels = c("Fine roots","Thick roots")) +
  theme(legend.position="none", font=c("white")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5)) + deicode_theme + theme(legend.key.size = unit(0.8, "cm"))
p2
```

**PERMANOVA of the loosely associated root microbiome in the fine roots:**  
```{r, cache=TRUE}
RA.dist <- RA.dist.L_F
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
adonis(formula = RA.dist ~ de$sample_time_cat * de$cluster1_cat + de$tree_cat, permutations = 999, strata = de$tree_cat)$aov.tab
```

**PERMANOVA of the loosely associated root microbiome in the thick roots:**  
```{r, cache=TRUE}
RA.dist <- RA.dist.L_T
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
adonis(formula = RA.dist ~ de$sample_time_cat * de$cluster1_cat + de$tree_cat, permutations = 999, strata = de$tree_cat)$aov.tab
```

**PERMANOVA of the tightly associated root microbiome in the fine roots:**  
```{r, cache=TRUE}
RA.dist <- RA.dist.T_F
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
adonis(formula = RA.dist ~ de$sample_time_cat * de$cluster1_cat + de$tree_cat, permutations = 999, strata = de$tree_cat)$aov.tab
```

**PERMANOVA of the tightly associated root microbiome in the thick roots:**  
```{r, cache=TRUE}
RA.dist <- RA.dist.T_T
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
adonis(formula = RA.dist ~ de$sample_time_cat * de$cluster1_cat + de$tree_cat, permutations = 999, strata = de$tree_cat)$aov.tab
```

### Manuscript plot:

The following part creates the combined plot shown in **figure 4 B** in the manuscript but is not executed here

```{r, eval=FALSE}
legend2 <- cowplot::get_legend(p1)
c1 <- ggpar(p1, ylim=c(-2.1,3), xlim=c(-1.9,1.9), legend = "", legend.title = "", font.x = c(22, "bold"), font.y = c(24, "bold"), font.tickslab = c(20, color="black")) +
  annotate(geom="text", x=1.2, y=2.9, label="L-compartment", color="black", size=8)
c2 <- ggpar(p2, ylim=c(-2.1,3), xlim=c(-1.9,1.9), legend = "", legend.title = "", font.x = c(22, "bold"), font.y = c(24, "bold"), font.tickslab = c(20, color="black")) +
  annotate(geom="text", x=1.2, y=2.9, label="T-compartment", color="black", size=8)
figure <- ggarrange(c1, c2, legend2, widths = c(2.5,2.5,1),
                    ncol = 3, nrow = 1)
png("CAP_DEICODE_combined_sampletime.png", width = 1400, height = 600)
print(figure)
dev.off()
```

<p>&nbsp;</p>

--------------------

### Differential abundance analyis:

Pull workspace from QIIME2 and additional data

This beginning workspace contains: 

* `ps`: phyloseq object including the metadata, and the ASV table, the taxonomy table and the phylogenetic tree, all three created by QIIME2
* `p.colors`: a list that codes the different phyla in different colors in the plot 


```{r,cache=TRUE}
load('spatio-temporal_trial_DA.RData')
```

The following creates text files which contain the statistical analyses
```{r, eval=FALSE}
# Set as a loop to do the analysis for both compartments
# Combination Rootpart/Timepoint: calculate ---------------------------------------------------------
for (comp in c("L","T")){
  
  pseq <- subset_samples(ps, compartment==comp) 
  genus_data <- pseq
  
  var1 <- "RTP"
  # TP4-F --------------------------------------------------------------------
  var2 <- "TP4-F"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                               c("TP4-F", "TP5-F", "TP6-F", "TP7-F", 
                                                 "TP4-T", "TP5-T", "TP6-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP4-F - TP5-F", "TP4-F - TP6-F", "TP4-F - TP7-F", 
               "TP4-F - TP4-T", "TP4-F - TP5-T", "TP4-F - TP6-T", "TP4-F - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))
  
  # TP5-F --------------------------------------------------------------------
  var2 <- "TP5-F"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                          c("TP5-F", "TP4-F", "TP6-F", "TP7-F", 
                                            "TP4-T", "TP5-T", "TP6-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP5-F - TP4-F", "TP5-F - TP6-F", "TP5-F - TP7-F", 
               "TP5-F - TP4-T", "TP5-F - TP5-T", "TP5-F - TP6-T", "TP5-F - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))
  
  # TP6-F --------------------------------------------------------------------
  var2 <- "TP6-F"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                          c("TP6-F", "TP4-F", "TP5-F", "TP7-F", 
                                            "TP4-T", "TP5-T", "TP6-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP6-F - TP4-F", "TP6-F - TP5-F", "TP6-F - TP7-F", 
               "TP6-F - TP4-T", "TP6-F - TP5-T", "TP6-F - TP6-T", "TP6-F - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))
  
  # TP7-F --------------------------------------------------------------------
  var2 <- "TP7-F"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                          c("TP7-F", "TP4-F", "TP5-F", "TP6-F", 
                                            "TP4-T", "TP5-T", "TP6-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP7-F - TP4-F", "TP7-F - TP5-F", "TP7-F - TP6-F", 
               "TP7-F - TP4-T", "TP7-F - TP5-T", "TP7-F - TP6-T", "TP7-F - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))
  
  # TP4-T --------------------------------------------------------------------
  var2 <- "TP4-T"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                          c("TP4-T", "TP4-F", "TP5-F", "TP6-F", "TP7-F", 
                                            "TP5-T", "TP6-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP4-T - TP4-F", "TP4-T - TP5-F", "TP4-T - TP6-F", 
               "TP4-T - TP7-F", "TP4-T - TP5-T", "TP4-T - TP6-T", "TP4-T - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))
  
  # TP5-T --------------------------------------------------------------------
  var2 <- "TP5-T"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                          c("TP5-T", "TP4-F", "TP5-F", "TP6-F", "TP7-F", 
                                            "TP4-T", "TP6-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP5-T - TP4-F", "TP5-T - TP5-F", "TP5-T - TP6-F", 
               "TP5-T - TP7-F", "TP5-T - TP4-T", "TP5-T - TP6-T", "TP5-T - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))
  
  # TP6-T --------------------------------------------------------------------
  var2 <- "TP6-T"
  sample_data(genus_data)$RTP <- factor(sample_data(genus_data)$RTP, levels = 
                                          c("TP6-T", "TP4-F", "TP5-F", "TP6-F", "TP7-F", 
                                            "TP4-T", "TP5-T", "TP7-T"))
  
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "RTP + tree_cat",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$beta
  colnames(tab_coef) 
  col_name = c("TP6-T - TP4-F", "TP6-T - TP5-F", "TP6-T - TP6-F", 
               "TP6-T - TP7-F", "TP6-T - TP4-T", "TP6-T - TP5-T", "TP6-T - TP7-T",
               "T1","T2","T3","T4","T5","T6","T7","T8","T9")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_w, by=0)[,-c(9:17,25:33)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))

}

# Combination Rootpart/Timepoint: combine datasets --------------------------------------------------------
for (comp in c("L","T")){
  
  pseq <- subset_samples(ps, compartment==comp) 
  genus_data <- pseq #tax_glom(pseq, taxrank <- rank_names(pseq)[6], NArm = FALSE)
  var1 <- "RTP"
  
  var2="TP4-F";TP4_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,2,5,9,12)]
  var2="TP5-F";TP5_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,3,6,10,13)]
  var2="TP6-F";TP6_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,4,7,11,14)]
  var2="TP7-F";TP7_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,8,15)]
  var2="TP4-T";TP4_T <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,6,13)]
  var2="TP5-T";TP5_T <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,7,14)]
  var2="TP6-T";TP6_T <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,8,15)]
  
  
  tmp <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(TP4_F,TP5_F,TP6_F,TP7_F,TP4_T,TP5_T,TP6_T))
  rownames(tmp) <- tmp$Row.names
  str(tmp)
  
  tmp3 <- as.data.frame(tax_table(pseq))
  tmp4 <- merge(tmp3, tmp, by=0) 
  rownames(tmp4) <- tmp4$Row.names
  tmp4$Row.names = NULL
  head(tmp4)
  
  PGroup <- transform_sample_counts(pseq, function(x)100* x / sum(x))
  OTUg <- otu_table(PGroup)
  TAXg <- tax_table(PGroup)
  AverageD <- as.data.frame(rowMeans(OTUg))
  names(AverageD) <- c("Mean")
  SD <- as.data.frame(rowSds(OTUg),na.rm = T)
  names(SD) <- c("SD")
  tmp <- cbind(AverageD,SD)
  GTable <- merge(TAXg, tmp, by=0, all=TRUE)
  rownames(GTable) <- GTable$Row.names
  GTable$Row.names = NULL
  head(GTable)
  
  tmp5 <- merge(tmp4[,-c(1,7:9)], GTable[,-c(1:8)], by=0) 
  row.names(tmp5) <- tmp5$Row.names
  tmp5$Row.names = NULL
  
  tmp7 <- tmp5[(rowSums(tmp5[,c(6,7,10,11,14,15,18,20,22,24)])!=0),]
  tmp7 <- tmp7[order(tmp7$Mean, decreasing = TRUE),]
  write_csv(tmp7, file = paste0("../stat_results/ANCOMBC/pairwise/",comp,"-",var1,"_results.csv"))
  
}



# Combination Rootpart/Timepoint: calculate ---------------------------------------------------------

for (comp in c("L","T")){
  
  pseq <- subset_samples(ps, compartment==comp) 
  genus_data <- pseq #tax_glom(pseq, taxrank <- rank_names(pseq)[6], NArm = FALSE)
  var1 <- "RTP"
  
  var2="TP4-F";TP4_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,2,5,9,12)]
  var2="TP5-F";TP5_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,3,6,10,13)]
  var2="TP6-F";TP6_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,4,7,11,14)]
  var2="TP7-F";TP7_F <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,8,15)]
  var2="TP4-T";TP4_T <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,6,13)]
  var2="TP5-T";TP5_T <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,7,14)]
  var2="TP6-T";TP6_T <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/ASV/",comp,"-",var2,".csv"))[,c(1,8,15)]
  
  
  tmp_M <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(TP4_F,TP5_F,TP6_F,TP7_F,TP4_T,TP5_T,TP6_T))
  rownames(tmp_M) <- tmp_M$Row.names
  tmp_M$Row.names = NULL
  str(tmp_M)
  
  pseq <- subset_samples(ps, compartment==comp) 
  PGroup <- transform_sample_counts(pseq, function(x)100* x / sum(x))
  OTUg <- otu_table(PGroup)
  TAXg <- tax_table(PGroup)
  AverageD <- as.data.frame(rowMeans(OTUg))
  names(AverageD) <- c("Mean")
  SD <- as.data.frame(rowSds(OTUg),na.rm = T)
  names(SD) <- c("SD")
  tmp <- cbind(AverageD,SD)
  GTable <- merge(TAXg, tmp, by=0, all=TRUE)
  rownames(GTable) <- GTable$Row.names
  GTable$Row.names = NULL
  GTable <- GTable[,c(9,10)]
  head(GTable)
  
  tmp5 <- merge(tmp_M, GTable, by=0) 
  row.names(tmp5) <- tmp5$Row.names
  tmp5$Row.names = NULL
  head(tmp5)
  write.table(tmp5, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"_results_RTP.csv"), sep="\t")
}


#Merge L and T
tmp_L <- read.table(file=paste0("../stat_results/ANCOMBC/pairwise/",tax,"/L_results_RTP.csv"))
colnames(tmp_L) <- paste("L", colnames(tmp_L), sep = " - ")
tmp_T <- read.table(file=paste0("../stat_results/ANCOMBC/pairwise/",tax,"/T_results_RTP.csv"))
colnames(tmp_T) <- paste("T", colnames(tmp_T), sep = " - ")
tmp <- merge(tmp_L, tmp_T, by=0, all= T) 
rownames(tmp) <- tmp$Row.names
tmp$Row.names <- NULL
head(tmp)

#add taxonomy
pseq <- ps
tmp3 <- as.data.frame(tax_table(pseq))[,-c(1,7,8)]
tmp4 <- merge(tmp3, tmp, by=0) 
rownames(tmp4) <- tmp4$Row.names
tmp4$Row.names = NULL
head(tmp4)
#View(tmp)

write_csv(tmp4, file = "../stat_results/ANCOMBC/pairwise/ASV/L_and_T_RTP.csv")

ancom <- tmp4[ which(tmp4$`L - Mean` >= 0.3 | tmp4$`T - Mean` >= 0.3), ]
colnames(ancom)[1:5] <- c("Phylum", "Class", "Order", "Family", "Genus")

ancom <- merge(ancom, p.colors, by="Phylum", all.y = T, )
ancom$Genus[ancom$Genus=="uncultured"] <- "Unclassified"
ancom$Genus[is.na(ancom$Genus)] <- "Unclassified"
ancom <- add_column(ancom, Name = ancom$Genus, .after = "Genus")

#chose only significant rows!
tmp <- ancom
colnames(tmp)
for (i in c(7,8,11,12,15,16,19,21,23,25,29,30,33,34,37,38,41,43,45,47)){
  tmp[,i][is.na(tmp[,i])] <- FALSE
}

tmp <- tmp[(rowSums(tmp[,c(7,8,11,12,15,16,19,21,23,25,29,30,33,34,37,38,41,43,45,47)])!=0),]
head(tmp)

exp_L <- tmp[(rowSums(tmp[,c(29,30,33,34,37,38,41,43,45,47)])!=0),]
exp_L <- exp_L[,c(1:28,51)]

exp_T <- tmp[(rowSums(tmp[,c(7,8,11,12,15,16,19,21,23,25)])!=0),]
exp_T <- exp_T[,c(1:6,29:51)]

tax.clean <- rename.ancombc.output(tmp)
exp_L <- rename.ancombc.output(exp_L)
exp_T <- rename.ancombc.output(exp_T)

write_csv(exp_L, file = "../stat_results/ANCOMBC/pairwise/ASV/L_RTP_significant.csv")
write_csv(exp_T, file = "../stat_results/ANCOMBC/pairwise/ASV/T_RTP_significant.csv")
write_csv(tax.clean, file = "../stat_results/ANCOMBC/pairwise/ASV/L_and_T_RTP_significant.csv")
```


### Differential abundance plot:

The latter part did the actual differential abundance analysis. Now we have to format the final table. It might be necessary to manually rename some of the ASVs, e.g. when there are two ASVs with the same name. Then, adding a a numbering is necessary. This will create the **figures S12** 

```{r, eval=FALSE}
# RTP plot L or T--------------------------------

tax <- "ASV"
var1 <- "RTP"

mat.df <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_",var1,"_significant.csv"), row.names = "Name")[,-55]
str(mat.df)

for (comp in c("L","T")){
  if(comp == "L") {
    tmp_L <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_",var1,"_significant.csv"), row.names = "Name")[,-55]
    tmp_L <- tmp_L[,c(1:23,46)]  
    tmp_L <- tmp_L[tmp_L$L...Mean>0.3,]
    tmp_L <- tmp_L[(rowSums(tmp_L[,c(2,3,6,7,10,11,14,16,18,20)])!=0),]
  } else if(comp == "T") {
    tmp_T <- read.csv(file = paste0("stat_results/ANCOMBC/pairwise/",tax,"/L_and_T_",var1,"_significant.csv"), row.names = "Name")[,-55]
    tmp_T <- tmp_T[,c(1,24:46)]  
    tmp_T <- tmp_T[tmp_T$T...Mean>0.3,]
    tmp_T <- tmp_T[(rowSums(tmp_T[,c(2,3,6,7,10,11,14,16,18,20)])!=0),]
  }
}

#chose one of the two for each plot
mat.df <- tmp_L; comp="L"
mat.df <- tmp_T; comp="T"

colnames(mat.df) <- c("Phylum",
                      "Fine: q-TP1:TP2", "TP1: q-F:T",
                      "Fine: TP1 -> TP2","TP1: Fine -> Thick",
                      "Fine: q-TP2:TP3", "TP2: q-F:T", 
                      "Fine: TP2 -> TP3","TP2: Fine -> Thick",
                      "Fine: q-TP3:TP4", "TP3: q-F:T", 
                      "Fine: TP3 -> TP4","TP3: Fine -> Thick",
                      "TP4: q-F:T", "TP4: Fine -> Thick",
                      "Thick: q-TP1:TP2", 
                      "Thick: TP1 -> TP2",
                      "Thick: q-TP2:TP3",
                      "Thick: TP2 -> TP3",
                      "Thick: q-TP3:TP4", 
                      "Thick: TP3 -> TP4",
                      "Mean", "SD",
                      "P-color")

#IMPORTANT: specify ONLY the columns with the differentials
col.order <- c("TP1: Fine -> Thick","TP2: Fine -> Thick",
               "TP3: Fine -> Thick","TP4: Fine -> Thick",
               "Fine: TP1 -> TP2", 
               "Fine: TP2 -> TP3",
               "Fine: TP3 -> TP4",
               "Thick: TP1 -> TP2",
               "Thick: TP2 -> TP3",
               "Thick: TP3 -> TP4")

nums <- unlist(lapply(mat.df, is.numeric))  
mat.diff <- mat.df[ , nums]
mat.diff <- as.matrix(mat.diff[,-c(11,12)])

nums <- unlist(lapply(mat.df, is.logical))
sig_mat <- as.matrix(mat.df[ , nums])
for (i in 1:ncol(sig_mat)){
  sig_mat[,i][is.na(sig_mat[,i])] <- FALSE
}

heatmap_function = function(num_matrix, omatrix , sig_matrix) {
  
  min_lc <- min(num_matrix, na.rm = T)
  max_lc <- max(num_matrix, na.rm = T)
  colors <- colorRamp2(c(min_lc, 0, max_lc), c("blue", "white", "red"))
  
  
  hb1 = rowAnnotation('Mean \nAbundance [%]' = anno_barplot(as.vector(omatrix$Mean), ylim = c(0,2), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1, axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 30)), width = unit(10, "cm")), annotation_name_gp = gpar(fontsize = 30, fontface = "bold"), annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  lgd = Legend(at = c(round(min_lc+min_lc*0.05,digits = 2), 0, round(max_lc+max_lc*0.05, digits = 2)),  col_fun = colors, 
               title = "W-value", title_gp = gpar(fontsize = 30, fontface = "bold"), title_gap = unit(0.5, "cm"),
               labels_gp = gpar(col = "black", fontsize = 30, fontface = "bold"), title_position = "topcenter", 
               grid_height = unit(3, "cm"), legend_width = unit(10, "cm"), direction = "horizontal") 
  
  ht = Heatmap(num_matrix , name = "logfold change",
               row_split = omatrix$Phylum,
               row_gap = unit(5, "mm"), 
               row_names_gp = gpar(fontsize = 34, fontface = "bold", 
                                   col = omatrix$`P-color`),
               na_col = "grey",
               col = colors,
               rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
               border = TRUE, 
               cluster_rows = FALSE, #remove cluster
               cluster_columns = FALSE,
               show_column_dend = FALSE, #remove cluster 
               show_heatmap_legend = FALSE, #remove legend
               row_names_side = "right",
               row_names_max_width = unit(2, "cm"),
               row_names_rot = 0,
               row_names_centered = FALSE,
               row_title_gp = gpar(fontsize = 28),
               row_title_rot = 0,
               column_title = paste0("ANCOM-BC \nW-value"),  
               column_title_gp = gpar(fill = "black", col = "white", fontsize = 32, fontface = "bold", border = "black"), 
               column_title_side = "top",
               column_names_max_height = unit(6, "cm"),
               column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
               column_names_rot = 90,
               column_names_centered = FALSE,
               cluster_column_slices = FALSE,
               column_order = col.order,
               column_gap = unit(5, "mm"), 
               show_parent_dend_line = FALSE, 
               cell_fun = function(j, i, x, y, width, height, filmat.dfl) {
                 if(sig_matrix[i, j] == "TRUE")
                   grid.text("*", x, y, gp = gpar(fontsize = 30, fontface = "bold"))
               }, 
               right_annotation = c(hb1))
  results <- c(fct = ht, lgd = lgd)
  return(results)

  mat.df$Phylum
  
  
  #all but proteobacteria
  j = 1; i = length(mat.df$Phylum)
  full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,], sig_matrix = sig_mat[j:i,])
  png(paste0("Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_",comp,"_0.3.png"), width = 2400, height = 400+45*((i-j)))
  draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) 
  draw(full_map$lgd, x = unit(55, "cm"), y = unit(6, "cm"))
  dev.off()
}
```


