---
title: "Spatial trial"
output: html_document
---


# Bacterial community analysis
### **Becker *et al.* "Spatio-temporal variation"**

-----------------

This document contains all statistical analyses conducted for the manuscript. Note that due to the random iterative nature of some analyses (such as PERMANOVA, CAP & PCA) some of the figure parameters will change slightly during reanalysis, though core results will remain essentially unchanged. 

All data to reproduce analysis can be found here: `https://github.com/mfbeuq/becker_etal_spatiotemporal`

-----------------

Load necessary ecological analysis libraries. 

```{r,cache=TRUE, message=FALSE}
library(phyloseq)
library(microbiome)
library(ggord)
library(metagMisc)
library(ggpubr)
library(FSA)
library(knitr)
library(rmarkdown)
library(ape)
library(vegan)
library(philr)
library(compositions)
library(qiime2R)
library(plyr)
library(dplyr)
library(tidyr)
library(PMCMR)
library(tibble)
library(viridis)
library(gridExtra)
library(AcidPlots)
library(grid)
library(colorRamps)
library(rstatix)
library(dunn.test)
library(pairwiseAdonis)
library(dplyr)
library(ANCOMBC)
library(nlme)
library(tidyverse)
library(compositions)
library(readr)
library(DT)
library(matrixStats)
library(pheatmap)
library(RColorBrewer)
library(dendsort)
library(ComplexHeatmap)
library(circlize)
library(round)

set.seed(225)
```


This beginning workspace contains: 

* `deicode_theme`: The plot theme for DEICODE plots
* `ps_fp`: phyloseq object including the metadata, and the ASV table, the taxonomy table and the phylogenetic tree, all three created by QIIME2
* `ps.meta`: Shannon diversity index data of the entire data table calculated by the core-metrics command in QIIME2
* `psB.meta`: Shannon diversity index data of the bulk soil sub data set
* `psL.meta`: Shannon diversity index data of the loosely associated (L) root microbiota sub data set
* `psT.meta`: Shannon diversity index data of the tightly associated (T) root microbiota sub data set
* `RA.dist.all`: DEICODE distance matrix generated by QIIME2 of the entire data table
* `RA.dist.B`: DEICODE distance matrix generated by QIIME2 of the bulk soil sub data set
* `RA.dist.L`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota sub data set
* `RA.dist.T`: DEICODE distance matrix generated by QIIME2 of the loosely associated (L) root microbiota sub data set
* `rename.ancombc.output`: function for renaming taxonomic levels in the differential abundance analysis.


```{r,cache=TRUE}
load('spatial_trial_diversity.RData')
```

Load functions neccessary:
```{r,cache=TRUE}
source('functions.R')
```

source('functions.R')  #functions for formatting shortcuts.

--------------------

### Alpha diversity statistics:

**Loosely associated root microbiome:**  
The mean Shannon diversity for the L-compartment is `r mean(psL.meta$Shannon)` with a standard deviation of `r sd(psL.meta$Shannon)`

The following calculates the Kruskal-Wallis test statistics for the L-compartment:

```{r, cache=TRUE}
psmeta.tib <- psL.meta %>%
  select(c("root_section","root_qdt","tree","Shannon")) %>%
  as_tibble() %>%
  pivot_longer(-c(root_section,root_qdt,tree), names_to = "variables", values_to = "value")
#run kw test
tmp1 <- psmeta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ root_section) %>%
  add_column(factor = "Root Section")
tmp2 <- psmeta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ root_qdt) %>%
  add_column(factor = "Root Quadrant")
tmp3 <- psmeta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ tree) %>%
  add_column(factor = "Tree")
dplyr::bind_rows(as_tibble(tmp1),as_tibble(tmp2),as_tibble(tmp3))
```

The following performs Dunn's test for multiple comparisons between the individual trees:

```{r, cache=TRUE, warning = FALSE}
dunn.test(psL.meta$Shannon , psL.meta$tree, method="bh", alpha=0.1, list=F)
```

<p>&nbsp;</p>
**Tightly associated root microbiome:**  
The mean Shannon diversity for the T-compartment is `r mean(psT.meta$Shannon)` with a standard deviation of `r sd(psT.meta$Shannon)`

The following calculates the Kruskal-Wallis test statistics for the T-compartment:

```{r, cache=TRUE}
psmeta.tib <- psT.meta %>%
  select(c("root_section","root_qdt","tree","Shannon")) %>%
  as_tibble() %>%
  pivot_longer(-c(root_section,root_qdt,tree), names_to = "variables", values_to = "value")
#run kw test
tmp1 <- psmeta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ root_section) %>%
  add_column(factor = "Root Section")
tmp2 <- psmeta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ root_qdt) %>%
  add_column(factor = "Root Quadrant")
tmp3 <- psmeta.tib %>%
  group_by(variables) %>%
  kruskal_test(value ~ tree) %>%
  add_column(factor = "Tree")
dplyr::bind_rows(as_tibble(tmp1),as_tibble(tmp2),as_tibble(tmp3))
```



The following performs Dunn's test for multiple comparisons between the four root size sections, the root quadrants and the individual trees:

```{r, cache=TRUE, cache=TRUE, warning = FALSE}
dunn.test(psT.meta$Shannon , psT.meta$root_section, method="bh", alpha=0.1, list=F)
dunn.test(psT.meta$Shannon , psT.meta$root_qdt, method="bh", alpha=0.1, list=F)
dunn.test(psT.meta$Shannon , psT.meta$tree, method="bh", alpha=0.1, list=F)
```


<p>&nbsp;</p>
**Bulk soil microbiome:**  
The mean Shannon diversity for the bulk soil is `r mean(psB.meta$Shannon)` with a standard deviation of `r sd(psB.meta$Shannon)`

Kruskal-Wallis test statistics for the individual trees:
```{r}
kruskal.test(psB.meta$Shannon ~ psB.meta$tree)
```

Kruskal-Wallis test statistics for the root quadrants:
```{r}
kruskal.test(psB.meta$Shannon ~ psB.meta$root_qdt)
```


### Alpha diversity plots:

This creates a each one Shannon diversity index plot of each the L- and T-compartment of the different root size section:

```{r, cache=TRUE}
# Plot the L-compartment 
a1 <- ggplot(psL.meta, aes(x=tree, y=Shannon, fill=root_section)) +
  geom_boxplot() + 
  scale_fill_manual(values=c("darkorchid","deepskyblue","darkorange","firebrick2"), name="Root Section") +
  theme(axis.title.x=element_blank(), axis.title.y=element_text(size=20), axis.text.y=element_text(size=16), axis.text.x=element_text(size=14))
print(a1)

# Plot the T-compartment 
a2 <- ggplot(psT.meta, aes(x=tree, y=Shannon, fill=root_section)) +
  geom_boxplot() + 
  scale_fill_manual(values=c("darkorchid","deepskyblue","darkorange","firebrick2"), name="Root Section") +
  theme(axis.title.x=element_blank(), axis.title.y=element_text(size=20), axis.text.y=element_text(size=16), axis.text.x=element_text(size=14))
print(a2)
```


--------------------

### Beta Diversity:

**Loosely associated root microbiome:**  
First, the distance matrix calculated by QIIME needs to be merged with the metadata file:
```{r, cache=TRUE}
RA.dist <- RA.dist.L #using a temporary variable
q_meta$root_section <- base::factor(q_meta$root_section, levels = c("< 1 mm","1-2 mm","2-4 mm","> 4 mm"))
row.names(q_meta) <- q_meta$SampleID #making sure both tables are in the right order
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)[,c(2:11,14)] #merging both tables
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree, Root Section and Quadrant as the constraining variables
```{r, cache=TRUE}
cap <- capscale(formula = RA.dist ~ de$root_section + de$root_qdt + de$tree, data = de, comm = RA.dist)
```

Furthermore, PERMANOVA is calculated on the distance matrix using the factor 'tree' as nesting factor
```{r, cache=TRUE}
adonis(formula = RA.dist ~ de$tree + de$root_section + de$tree:de$root_qdt, permutations = 999, strata = de$tree)$aov.tab
```

In addition, pairwise PERMANOVA is calculated for the factors root section
```{r, cache=TRUE}
pairwise.adonis(x=RA.dist, factors=de[,c(8)], perm = 999) 
```
and tree individual
```{r, cache=TRUE}
pairwise.adonis(x=RA.dist, factors=de[,c(6)], perm = 999) 
```

The CAP plot for the L-compartment by using the the first two axes of CAP
```{r, cache=TRUE}
x <- as.data.frame(scores(cap, display = "sites")) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID
p1 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = root_section, shape = tree)) +
  geom_point(alpha=0.8, size=8) +
  scale_color_manual(values=c("darkorchid","deepskyblue","darkorange","firebrick2"), name="Root Section",
                     labels = c("< 1 mm","1-2 mm","2-4 mm","> 4 mm")) +
  scale_shape_manual(values=c(15,16,17,18), name="Tree", 
                     labels = c("Tree 1","Tree 2","Tree 3","Tree 4")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_y_continuous(labels = scales::number_format(accuracy = 0.1, decimal.mark = '.'))
print(p1)
```


<p>&nbsp;</p>

**Tightly associated root microbiome:**  
First, the distance matrix calculated by QIIME needs to be merged with the metadata file:
```{r, cache=TRUE}
RA.dist <- RA.dist.T #using a temporary variable
q_meta$root_section <- base::factor(q_meta$root_section, levels = c("< 1 mm","1-2 mm","2-4 mm","> 4 mm"))
row.names(q_meta) <- q_meta$SampleID #making sure both tables are in the right order
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)[,c(2:11,14)] #merging both tables
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree, Root Section and Quadrant as the constraining variables
```{r, cache=TRUE}
cap <- capscale(formula = RA.dist ~ de$root_section + de$root_qdt + de$tree, data = de, comm = RA.dist)
```

Furthermore, PERMANOVA is calculated on the distance matrix using the factor 'tree' as nesting factor
```{r, cache=TRUE}
adonis(formula = RA.dist ~ de$tree + de$root_section + de$tree:de$root_qdt, permutations = 999, strata = de$tree)$aov.tab
```

In addition, pairwise PERMANOVA is calculated for the factors root section
```{r, cache=TRUE}
pairwise.adonis(x=RA.dist, factors=de[,c(8)], perm = 999) 
```
and tree individual
```{r, cache=TRUE}
pairwise.adonis(x=RA.dist, factors=de[,c(6)], perm = 999) 
```

The CAP plot for the T-compartment by using the the first two axes of CAP
```{r, cache=TRUE}
x <- as.data.frame(scores(cap, display = "sites")) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID
p2 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = root_section, shape = tree)) +
  geom_point(alpha=0.8, size=8) +
  scale_color_manual(values=c("darkorchid","deepskyblue","darkorange","firebrick2"), name="Root Section",
                     labels = c("< 1 mm","1-2 mm","2-4 mm","> 4 mm")) +
  scale_shape_manual(values=c(15,16,17,18), name="Tree", 
                     labels = c("Tree 1","Tree 2","Tree 3","Tree 4")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_y_continuous(labels = scales::number_format(accuracy = 0.1, decimal.mark = '.'))
print(p2)
```



### Manuscript plot:

The following part creates the combined plot shown in figure 2 A and B in the manuscript but is not executed here

```{r, eval=FALSE}
legend1 <- cowplot::get_legend(a1)
c1 <- ggpar(a1, legend = "", legend.title = "", ylim = c(6.7,9.2), font.x = c(0, "bold"), font.y = c(24, "bold"), font.tickslab = c(20, color="black"))  +
  annotate(geom="text", x=3.9, y=9.25, label="L-compartment", color="black", size=8)
c2 <- ggpar(a2, legend = "", legend.title = "", ylim = c(6.7,9.2), font.x = c(0, "bold"), font.y = c(24, "bold"), font.tickslab = c(20, color="black"))  +
  annotate(geom="text", x=3.9, y=9.25, label="T-compartment", color="black", size=8)

legend2 <- cowplot::get_legend(p1)
d1 <- ggpar(p1, ylim=c(-1.8,1.8), xlim=c(-0.8,1), legend = "", legend.title = "", font.x = c(22, "bold"), font.y = c(24, "bold"), font.tickslab = c(20, color="black")) +
  annotate(geom="text", x=0.75, y=1.8, label="L-compartment", color="black", size=8)
d2 <- ggpar(p2, ylim=c(-1.8,1.8), xlim=c(-0.8,1), legend = "", legend.title = "", font.x = c(22, "bold"), font.y = c(24, "bold"), font.tickslab = c(20, color="black")) +
  annotate(geom="text", x=0.75, y=1.8, label="T-compartment", color="black", size=8)

figure <- ggarrange(c1, c2, legend1, NULL,NULL,NULL, d1, d2, legend2, 
                    widths = c(2.5,2.5,1), heights = c(1, 0.05, 1),
                    ncol = 3, nrow = 3)
png("alpha_beta_combined.png", width = 1400, height = 1200)
print(figure)
dev.off()
```


<p>&nbsp;</p>
**Bulk soil microbiome:**  
First, the statistical analysis needs to be conducted and then the file merged:
```{r, cache=TRUE, eval=FALSE}
RA.dist <- RA.dist.B
q_meta <- read_q2metadata("../scripts_maps/map_file_v4a_spat.tsv") 
q_meta$root_section <- base::factor(q_meta$root_section, levels = c("B","<1mm","1-2mm","2-4mm",">4mm"))
row.names(q_meta) <- q_meta$SampleID
de <- merge(q_meta, RA.dist, by=0, all=F, all.y=F)[,c(2:11)]
row.names(de) <- de$SampleID
de <- de[ order(row.names(de)), ]
RA.dist <- RA.dist[ order(row.names(RA.dist)), ]
cap <- capscale(formula = RA.dist ~ de$root_qdt + de$tree, data = de, comm = RA.dist)

#PERMANOVA
adonis2(formula = RA.dist ~ de$root_qdt + de$tree, permutations = 999)

#Plot
x <- as.data.frame(scores(cap, display = "sites")) 
tmp <- merge(q_meta, x, by=0)
row.names(tmp) <- tmp$SampleID
p3 <- ggplot(tmp, aes(x= CAP1, y= CAP2, color = root_qdt, shape = tree)) +
  geom_point(alpha=0.8, size=12) +
  scale_color_manual(values=c("darkorchid","deepskyblue","darkorange","firebrick2"), name="Root Quadrant",
                     labels = c("Quadrant 1","Quadrant 2","Quadrant 3","Quadrant 4")) +
  scale_shape_manual(values=c(15,16,17,18), name="Tree", 
                     labels = c("Tree 1","Tree 2","Tree 3","Tree 4")) +
  geom_hline(yintercept = 0, linetype="dotted") + 
  geom_vline(xintercept = 0, linetype="dotted") +
  xlab(paste("CAP1 - ",round(100*summary(cap)$cont$importance[2,1],digits = 1),"%")) +
  ylab(paste("CAP2 - ",round(100*summary(cap)$cont$importance[2,2],digits = 1),"%")) +
  theme(plot.title = element_text(hjust = 0.5))  + deicode_theme  +
  scale_y_continuous(labels = scales::number_format(accuracy = 0.1,
                                                    decimal.mark = '.'))
ggpar(p3, font.x = c(20, "bold"), font.y = c(20, "bold"), font.tickslab = c(18, color="black")) +
  annotate(geom="text", x=-1.1, y=1.6, label="Bulk soil", color="black", size=12)
```

Then CAP is calculated by using the 'capscale' command in R and using the factors Tree, Root Section and Quadrant as the constraining variables
```{r, cache=TRUE}

```


--------------------

### Differential abundance analyis:

Pull workspace from QIIME2 and additional data

This beginning workspace contains: 

* `ps_fp`: phyloseq object including the metadata, and the ASV table, the taxonomy table and the phylogenetic tree, all three created by QIIME2
* `p.colors`: a list that codes the different phyla in different colors in the plot 


```{r,cache=TRUE}
load('spatial_trial_SD.RData')
```

The following creates text files which contain the statistical analyses
```{r, eval=FALSE}
# Set as a loop to do the analysis for both compartments
for (comp in c("L","T")){
  
  #read in the phyloseq object and filter for samples from the corresponding compartment
  pseq <- subset_samples(ps, compartment==comp) 
  genus_data <- pseq 
  #set variable of interest
  var1 <- "root_section"
  #set taxonomic level
  tax <- "ASV"
  
  # S1 --------------------------------------------------------------------
  var2 <- "S1"
  #set the order of factors and change them to get evera pairwise comparison
  sample_data(genus_data)$root_section <- factor(sample_data(genus_data)$root_section, levels = 
                                                   c("< 1 mm", "1-2 mm", "2-4 mm", "> 4 mm"))
  sample_data(genus_data)$tree <- factor(sample_data(genus_data)$tree, levels = 
                                                   c("T1", "T2", "T3", "T4"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "tree + root_section",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #get Coefficients and change the column names of each sub data frame 
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("T1 - T2", "T1 - T3", "T1 - T4", 
               "R1 - R2", "R1 - R3", "R1 - R4")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)

  #only filter the test statistics and the coefficients  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_coef, by=0) 
  #export them to a text file!
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  
  # S2 --------------------------------------------------------------------
  #repeat the first part but change the order of the factors
  var2 <- "S2"
  genus_data <- pseq #tax_glom(pseq, taxrank <- rank_names(pseq)[6], NArm = FALSE)
  sample_data(genus_data)$root_section <- factor(sample_data(genus_data)$root_section, levels = 
                                                   c("1-2 mm", "< 1 mm", "2-4 mm", "> 4 mm"))
  sample_data(genus_data)$tree <- factor(sample_data(genus_data)$tree, levels = 
                                           c("T2", "T1", "T3", "T4"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "tree + root_section",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("T2 - T1", "T2 - T3", "T2 - T4", 
               "R2 - R1", "R2 - R3", "R2 - R4")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_coef, by=0)[,-c(2,5,8,11)]
  head(tmp2)
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
  
  # S3 --------------------------------------------------------------------
  #repeat the first part but change the order of the factors
  var2 <- "S3"
  genus_data <- pseq #tax_glom(pseq, taxrank <- rank_names(pseq)[6], NArm = FALSE)
  sample_data(genus_data)$root_section <- factor(sample_data(genus_data)$root_section, levels = 
                                                   c( "2-4 mm", "< 1 mm", "1-2 mm","> 4 mm"))
  sample_data(genus_data)$tree <- factor(sample_data(genus_data)$tree, levels = 
                                           c( "T3", "T1", "T2","T4"))
  
  # Run ancombc function
  out = ancombc(phyloseq = genus_data, formula = "tree + root_section",
                p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
                group = var1, struc_zero = TRUE, neg_lb = FALSE,
                tol = 1e-5, max_iter = 100, conserve = TRUE,
                alpha = 0.05, global = FALSE)
  res = out$res
  
  #Coefficients
  tab_coef = res$W
  colnames(tab_coef) 
  col_name = c("T3 - T1", "T3 - T2", "T3 - T4", 
               "R3 - R1", "R3 - R2", "R3 - R4")
  colnames(tab_coef) = col_name
  
  source("ancom_bc_rename_variables.R", echo = T, spaced = T)
  
  tmp <- tab_diff
  tmp2 <- merge(tmp, tab_coef, by=0)[,c(1,4,7,10,13)]
  write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
}


#the same for bulk soil:
comp="B"
pseq <- subset_samples(ps, compartment==comp) 
genus_data <- pseq 
var1 <- "tree"

###
var2 <- "B1"
sample_data(genus_data)$tree <- factor(sample_data(genus_data)$tree, levels = 
                                         c("T1", "T2", "T3", "T4"))

# Run ancombc function
out = ancombc(phyloseq = genus_data, formula = "tree + root_qdt",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
              group = var1, struc_zero = TRUE, neg_lb = FALSE,
              tol = 1e-5, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = FALSE)
res = out$res

#Coefficients
tab_coef = res$W
colnames(tab_coef) 
col_name = c("T1 - T2", "T1 - T3", "T1 - T4", 
             "Q1 - Q2", "Q1 - Q3", "Q1 - R4")
colnames(tab_coef) = col_name

source("ancom_bc_rename_variables.R", echo = T, spaced = T)
tmp <- tab_diff
tmp2 <- merge(tmp, tab_coef, by=0) 
write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))

###
var2 <- "B2"
genus_data <- pseq
sample_data(genus_data)$tree <- factor(sample_data(genus_data)$tree, levels = 
                                         c("T2", "T1", "T3", "T4"))

# Run ancombc function
out = ancombc(phyloseq = genus_data, formula = "tree + root_qdt",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
              group = var1, struc_zero = TRUE, neg_lb = FALSE,
              tol = 1e-5, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = FALSE)
res = out$res

#Coefficients
tab_coef = res$W
colnames(tab_coef) 
col_name = c("T2 - T1", "T2 - T3", "T2 - T4", 
             "Q1 - Q2", "Q1 - Q3", "Q1 - R4")
colnames(tab_coef) = col_name

source("ancom_bc_rename_variables.R", echo = T, spaced = T)
tmp <- tab_diff
tmp2 <- merge(tmp, tab_coef, by=0) 
write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
###
var2 <- "B3"
sample_data(genus_data)$tree <- factor(sample_data(genus_data)$tree, levels = 
                                         c("T3", "T1", "T2", "T4"))

# Run ancombc function
out = ancombc(phyloseq = genus_data, formula = "tree + root_qdt",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 10000,
              group = var1, struc_zero = TRUE, neg_lb = FALSE,
              tol = 1e-5, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = FALSE)
res = out$res

#Coefficients
tab_coef = res$W
colnames(tab_coef) 
col_name = c("T3 - T1", "T3 - T2", "T3 - T4", 
             "Q1 - Q2", "Q1 - Q3", "Q1 - R4")
colnames(tab_coef) = col_name

source("ancom_bc_rename_variables.R", echo = T, spaced = T)
tmp <- tab_diff
tmp2 <- merge(tmp, tab_coef, by=0) 
write_csv(tmp2, file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
###

# Now the loop has created three text files, each containing one pairwise comparison and the three for each compartment have to be combined together to a single temporary file

# combine bulk datasets --------------------------------------------------------

comp="B"; tax="ASV"
var2="B1";B1 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
var2="B2";B2 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
var2="B3";B3 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))

tmp <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(B1,B2,B3))
rownames(tmp) <- tmp$Row.names
tmp <- tmp[,c("T1...T2.x","T1...T3.x","T1...T4.x","T2...T3.x","T2...T4.x","T3...T4.x",
              "T1...T2.y","T1...T3.y","T1...T4.y","T2...T3.y","T2...T4.y","T3...T4.y")]
View(tmp)

pseq <- subset_samples(ps, compartment=="B") 
tmp3 <- as.data.frame(tax_table(pseq))
tmp4 <- merge(tmp3, tmp, by=0) 
rownames(tmp4) <- tmp4$Row.names
tmp4$Row.names = NULL
head(tmp4)

PGroup <- transform_sample_counts(pseq, function(x)100* x / sum(x))
OTUg <- otu_table(PGroup)
TAXg <- tax_table(PGroup)
AverageD <- as.data.frame(rowMeans(OTUg))
names(AverageD) <- c("Mean")
SD <- as.data.frame(rowSds(OTUg),na.rm = T)
names(SD) <- c("SD")
tmp <- cbind(AverageD,SD)
GTable <- merge(TAXg, tmp, by=0, all=TRUE)
rownames(GTable) <- GTable$Row.names
GTable$Row.names = NULL
GTable <- GTable[,-c(1,7)]
head(GTable)

tmp5 <- merge(tmp4, GTable, by=0) 
row.names(tmp5) <- tmp5$Row.names
tmp5$Row.names = NULL
tmp7 <- tmp5[-c(1,7,20:24)]
tmp7 <- tmp7[(rowSums(tmp7[,c(6:11)])!=0),]
tmp7 <- tmp7[order(tmp7$Mean, decreasing = TRUE),]
write_csv(tmp7, file = "../stat_results/ANCOMBC/pairwise/B_results_tree_significant.csv")


# combine L- and T-compartment datasets --------------------------------------------------------

#read the three text files of the L-compartment
comp="L"; tax="ASV"
var2="S1";L1 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
var2="S2";L2 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
var2="S3";L3 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
#merge them
tmp_L <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(L1,L2,L3))
rownames(tmp_L) <- tmp_L$Row.names
tmp_L$Row.names <- NULL

#read the three text files of the T-compartment
comp="T"; tax="ASV"
var2="S1";T1 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
var2="S2";T2 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
var2="S3";T3 <- read.csv(file = paste0("../stat_results/ANCOMBC/pairwise/",tax,"/",comp,"-",var2,".csv"))
#merge them
tmp_T <- Reduce(function(x, y) merge(x, y, all=TRUE, by="Row.names"), list(T1,T2,T3))
rownames(tmp_T) <- tmp_T$Row.names
tmp_T$Row.names <- NULL


#now the combine both L- and T-compartment file and add additional information
tmp <- merge(tmp_L, tmp_T, by=0, all= T) 
rownames(tmp) <- tmp$Row.names
tmp$Row.names <- NULL

#read in the phyloseq object and merge the taxonomy table with the statistical analysis
pseq <- ps
tmp3 <- as.data.frame(tax_table(pseq))
tmp4 <- merge(tmp3, tmp, by=0) 
rownames(tmp4) <- tmp4$Row.names
tmp4$Row.names = NULL

#filter out ASVs which have a relative abundance of less then 0.1% in either compartment
ancom <- tmp4[ which(tmp$`L - Mean` >= 0.1 | tmp$`T - Mean` >= 0.1), ]
ancom <- ancom[,-c(1,7)]

#add a color for each ASV belonging to a phylum
ancom <- merge(ancom, p.colors, by="Phylum", all.y = T, )
#rename files
ancom$Genus[ancom$Genus=="uncultured"] <- "Unclassified"
ancom$Genus[is.na(ancom$Genus)] <- "Unclassified"
#add a new column to specify names
ancom <- add_column(ancom, Name = ancom$Genus, .after = "Genus")
#chose only significant ASVs that showed a significant differential abundance!
tmp <- ancom
for (i in c(10:12,21:22,28,36:38,47:48,54)){
  tmp[,i][is.na(tmp[,i])] <- FALSE
}

tmp <- tmp[(rowSums(tmp[,c(10:12,21:22,28,36:38,47:48,54)])!=0),]
ancom <- tmp[,c(1:6,10:12,16:18,21:22,25:26,28,30:32,36:38,42:44,47:48,51:52,54,56:59)]

#rename the ASVs
tax.clean <- tmp
tax.cleanLT <- rename.ancombc.output(tax.clean)


# combine L/T/B datasets --------------------------------------------------------
#merge the three data files
tmp <- merge(tmp_L, tmp_T, by=0, all = T) 
rownames(tmp) <- tmp$Row.names
tmp$Row.names <- NULL
tmp <- merge(tmp, tmp_B, by=0, all = T) 
rownames(tmp) <- tmp$Row.names
tmp$Row.names <- NULL

#add taxonomy
tmp3 <- as.data.frame(tax_table(ps))
tmp4 <- merge(tmp3, tmp, by=0) 
rownames(tmp4) <- tmp4$Row.names
tmp4$Row.names = NULL

#filter non-significant ASVs with a relative abundance <0.1 %
ancom <- tmp4[ which(tmp4$`L - Mean` >= 0.1 | tmp4$`T - Mean` >= 0.1 | tmp4$`B - Mean` >= 0.1), -c(1,7)]
as.vector(names(ancom))
ancom <- ancom[,-c(9:11,15:17,20:21,24:25,27,29,35:37,41:43,46:47,50:51,53,55)]

#add phyla colors
ancom <- merge(ancom, p.colors, by="Phylum", all.y = T, )
ancom$Genus[ancom$Genus=="uncultured"] <- "Unclassified"
ancom$Genus[is.na(ancom$Genus)] <- "Unclassified"
ancom <- add_column(ancom, Name = ancom$Genus, .after = "Genus")

#chose only significant rows!
tmp <- ancom
for (i in c(7:9,13:14,17,21:23,27:28,31,35:40)){
  tmp[,i][is.na(tmp[,i])] <- FALSE
}

tmp <- tmp[(rowSums(tmp[,c(7:9,13:14,17,21:23,27:28,31,35:40)])!=0),]

#rename the ASVs
tax.clean <- tmp
tax.clean_LTB <- rename.ancombc.output(tax.clean)
```

### Differential abundance plot:

The latter part did the actual differential abundance analysis. Now we have to format the final table. It might be neccessary to manually rename some of the ASVs, e.g. when there are two ASVs with the same name. Then, adding a a numbering is neccessary. 

**Figure 3**  
```{r, eval=FALSE}
mat.df <- tax.clean_LT

#rename the columns
colnames(mat.df) <- c("Phylum", 
                      "L: q-R1:R2","L: q-R1:R3","L: q-R1:R4", 
                      "L: < 1 mm -> 1-2 mm", "L: < 1 mm -> 2-4 mm", "L: < 1 mm -> > 4 mm", 
                      "L: q-R2:R3","L: q-R2:R4", 
                      "L: 1-2 mm -> 2-4 mm", "L: 1-2 mm -> > 4 mm",
                      "L: q-R3:R4", "L: 2-4 mm -> > 4 mm",
                      "L: Mean", "L: SD",
                      "T: q-R1:R2","T: q-R1:R3","T: q-R1:R4", 
                      "T: < 1 mm -> 1-2 mm", "T: < 1 mm -> 2-4 mm", "T: < 1 mm -> > 4 mm", 
                      "T: q-R2:R3","T: q-R2:R4", 
                      "T: 1-2 mm -> 2-4 mm", "T: 1-2 mm -> > 4 mm",
                      "T: q-R3:R4","T: 2-4 mm -> > 4 mm",
                      "T: Mean", "T: SD", "P.color")


#IMPORTANT: specify ONLY the columns with the differentials
col.order <- c("L: < 1 mm -> 1-2 mm", "L: 1-2 mm -> 2-4 mm", "L: 2-4 mm -> > 4 mm",
               "L: < 1 mm -> 2-4 mm", "L: < 1 mm -> > 4 mm", "L: 1-2 mm -> > 4 mm",
               "T: < 1 mm -> 1-2 mm", "T: 1-2 mm -> 2-4 mm", "T: 2-4 mm -> > 4 mm",
               "T: < 1 mm -> 2-4 mm", "T: < 1 mm -> > 4 mm", "T: 1-2 mm -> > 4 mm"
               )

mat.diff <- as.matrix(mat.df[,c(5:7,10:11,13,19:21,24:25,27)])
sig_mat <- as.matrix(mat.df[,c(2:4,8:9,12,16:18,22:23,26)])

#define function to split the plot later
heatmap_function = function(num_matrix, omatrix , sig_matrix) {
  
  #get colors for the legend (minimum and maximum coefficient values)
  min_lc <- min(num_matrix, na.rm = T)
  max_lc <- max(num_matrix, na.rm = T)
  colors <- colorRamp2(c(min_lc, 0, max_lc), c("blue", "white", "red"))
  
  #create barplot for the mean relative abundance of ASVs in the L-compartment
  hb1 = rowAnnotation('L: Mean \nAbundance [%]' = anno_barplot(as.vector(omatrix$`L: Mean`), ylim = c(0,4), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1, axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 30)), width = unit(10, "cm")), annotation_name_gp = gpar(fontsize = 30, fontface = "bold"), annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  #create barplot for the mean relative abundance of ASVs in the T-compartment
  hb2 = rowAnnotation('T: Mean \nAbundance [%]' = anno_barplot(as.vector(omatrix$`T: Mean`), ylim = c(0,4), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1, axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 30)), width = unit(10, "cm")), annotation_name_gp = gpar(fontsize = 30, fontface = "bold"), annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)

  #create a legend with the previously created colors  
  lgd = Legend(at = c(round(min_lc+min_lc*0.05,digits = 2), 0, round(max_lc+max_lc*0.05, digits = 2)), col_fun = colors, title = "W-value", title_gp = gpar(fontsize = 30, fontface = "bold"), title_gap = unit(0.5, "cm"), labels_gp = gpar(col = "black", fontsize = 30, fontface = "bold"), title_position = "topcenter", grid_height = unit(3, "cm"), legend_width = unit(10, "cm"), direction = "horizontal") 
  
  #set parameters for the heatmap
  ht = Heatmap(num_matrix , name = "logfold change",
                                    row_split = omatrix$Phylum,
                                    row_gap = unit(5, "mm"), 
                                    row_names_gp = gpar(fontsize = 34, fontface = "bold", 
                                                        col = omatrix$P.color),
                                    na_col = "grey",
                                    col = colors,
                                    rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
                                    border = TRUE, 
                                    cluster_rows = FALSE, #remove cluster         
                                    show_column_dend = FALSE, #remove cluster 
                                    show_heatmap_legend = FALSE, #remove legend
                                    row_names_side = "right",
                                    row_names_max_width = unit(2, "cm"),
                                    row_names_rot = 0,
                                    row_names_centered = FALSE,
                                    row_title_gp = gpar(fontsize = 28),
                                    row_title_rot = 0,
                                    #row_labels = row_labels[rownames(mat.df)], 
                                    column_title = paste0("ANCOM-BC \nW-value"),  
                                    column_title_gp = gpar(fill = "black", col = "white", fontsize = 32, fontface = "bold", border = "black"), 
                                    column_title_side = "top",
                                    column_names_max_height = unit(6, "cm"),
                                    column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
                                    column_names_rot = 90,
                                    column_names_centered = TRUE,
                                    column_order = col.order,
                                    column_split = c(rep("L",6),rep("T",6)),
                                    column_gap = unit(5, "mm"), 
                                    show_parent_dend_line = FALSE, 
                                    #add "*" to indicate significance 
                                    cell_fun = function(j, i, x, y, width, height, filmat.dfl) {
                                      if(sig_matrix[i, j] == "TRUE")
                                        grid.text("*", x, y, gp = gpar(fontsize = 30, fontface = "bold"))
                                    }, 
                                    right_annotation = c(hb1,hb2))
  results <- c(fct = ht, lgd = lgd)
  return(results)
}


#heatmap of all ASVs but proteobacteria
j = 1; i = 52
full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,] ,sig_matrix = sig_mat[j:i,])
png(paste0("../Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_LT_1.png"), width = 2400, height = 300+45*((i-j)))
draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) # add space for titles
draw(full_map$lgd, x = unit(45, "cm"), y = unit(6, "cm"))
dev.off()

#create the second part of the heatmap 
j = 53; i = 123
full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,] ,sig_matrix = sig_mat[j:i,])
png(paste0("../Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_LT_2.png"), width = 2400, height = 300+45*((i-j)))
draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) # add space for titles
draw(full_map$lgd, x = unit(45, "cm"), y = unit(6, "cm"))
dev.off()
```

**Figure S7**  
```{r, eval=FALSE}
# LTB Tree --------------------------------
###ANCOM results
#prior to this step, redo list in excel!
mat.df <- tax.clean_LTB

colnames_vector <- c("Phylum", 
                  "L: q-T1:T2","L: q-T1:T3","L: q-T1:T4", 
                  "L: T1 - T2","L: T1 - T3","L: T1 - T4", 
                  "L: q-T2:T3","L: q-T2:T4",
                  "L: T2 - T3","L: T2 - T4",
                  "L: q-T3:T4","L: T3 - T4",
                  "L: Mean", "L: SD",
                  "T: q-T1:T2","T: q-T1:T3","T: q-T1:T4", 
                  "T: T1 - T2","T: T1 - T3","T: T1 - T4", 
                  "T: q-T2:T3","T: q-T2:T4",
                  "T: T2 - T3","T: T2 - T4",
                  "T: q-T3:T4","T: T3 - T4",
                  "T: Mean", "T: SD",                  
                  "B: q-T1:T2","B: q-T1:T3","B: q-T1:T4","B: q-T2:T3","B: q-T2:T4","B: q-T3:T4",
                  "B: T1 - T2","B: T1 - T3","B: T1 - T4","B: T2 - T3","B: T2 - T4","B: T3 - T4",
                  "B: Mean", "B: SD",
                  "P.color")

colnames(mat.df) <- colnames_vector

#IMPORTANT: specify ONLY the columns with the differentials
col.order <- c("L: T1 - T2","L: T1 - T3","L: T1 - T4","L: T2 - T3","L: T2 - T4","L: T3 - T4",
               "T: T1 - T2","T: T1 - T3","T: T1 - T4","T: T2 - T3","T: T2 - T4","T: T3 - T4",
               "B: T1 - T2","B: T1 - T3","B: T1 - T4","B: T2 - T3","B: T2 - T4","B: T3 - T4"
               )
for (i in c(2:4,8:9,12,16:18,22:23,26,30:35)){
  mat.df[,i][is.na(mat.df[,i])] <- FALSE
}
mat.df <- mat.df[(rowSums(mat.df[,c(2:4,8:9,12,16:18,22:23,26,30:35)])!=0),]

colnames(mat.df) <- colnames_vector
mat.diff <- as.matrix(mat.df[,c(5:7,10:11,13,19:21,24:25,27,36:41)])
sig_mat <- as.matrix(mat.df[,c(2:4,8:9,12,16:18,22:23,26,30:35)])


#define function
heatmap_function = function(num_matrix, omatrix , sig_matrix) {
  
  min_lc <- min(num_matrix, na.rm = T)
  max_lc <- max(num_matrix, na.rm = T)
  colors <- colorRamp2(c(min_lc, 0, max_lc), c("blue", "white", "red"))
  
  
  hb1 = rowAnnotation('L: Mean \nAbundance [%]' = anno_barplot(as.vector(omatrix$`L: Mean`), ylim = c(0,3), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1, axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 30)), width = unit(8, "cm")), annotation_name_gp = gpar(fontsize = 30, fontface = "bold"), annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  hb2 = rowAnnotation('T: Mean \nAbundance [%]' = anno_barplot(as.vector(omatrix$`T: Mean`), ylim = c(0,3), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1, axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 30)), width = unit(8, "cm")), annotation_name_gp = gpar(fontsize = 30, fontface = "bold"), annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  hb3 = rowAnnotation('B: Mean \nAbundance [%]' = anno_barplot(as.vector(omatrix$`B: Mean`), ylim = c(0,3), gp = gpar(fill = "black"), bar_width = 0.9, extend=0.1, axis_param = list(side = "bottom", labels_rot = 90, gp = gpar(fontsize = 30)), width = unit(8, "cm")), annotation_name_gp = gpar(fontsize = 30, fontface = "bold"), annotation_name_side = "top", annotation_name_offset = unit(0.3, "cm"), annotation_name_rot = 0)
  
  lgd = Legend(at = c(round(min_lc+min_lc*0.05,digits = 2), 0, round(max_lc+max_lc*0.05, digits = 2)),  col_fun = colors, title = "W-value", title_gp = gpar(fontsize = 30, fontface = "bold"), title_gap = unit(0.5, "cm"), labels_gp = gpar(col = "black", fontsize = 30, fontface = "bold"), title_position = "topcenter", grid_height = unit(3, "cm"), legend_width = unit(10, "cm"), direction = "horizontal") 
  
  ht = Heatmap(num_matrix , name = "logfold change",
               row_split = omatrix$Phylum,
               row_gap = unit(5, "mm"), 
               row_names_gp = gpar(fontsize = 34, fontface = "bold", 
                                   col = omatrix$P.color),
               na_col = "grey",
               col = colors,
               rect_gp = gpar(col = "white", lwd = 2), #white spaces in between
               border = TRUE, 
               cluster_rows = FALSE, #remove cluster         
               show_column_dend = FALSE, #remove cluster 
               show_heatmap_legend = FALSE, #remove legend
               row_names_side = "right",
               row_names_max_width = unit(2, "cm"),
               row_names_rot = 0,
               row_names_centered = FALSE,
               row_title_gp = gpar(fontsize = 28),
               row_title_rot = 0,
               column_title = paste0("ANCOM-BC \nW-value"),  
               column_title_gp = gpar(fill = "black", col = "white", fontsize = 32, fontface = "bold", border = "black"), 
               column_title_side = "top",
               column_names_max_height = unit(6, "cm"),
               column_names_gp = gpar(fontsize = 30, fontface = "bold", col = "black"),
               column_names_rot = 90,
               column_names_centered = TRUE,
               column_order = col.order,
               column_split = c(rep("B",6),rep("L",6),rep("T",6)),
               column_gap = unit(5, "mm"), 
               show_parent_dend_line = FALSE, 
               cell_fun = function(j, i, x, y, width, height, filmat.dfl) {
                 if(sig_matrix[i, j] == "TRUE")
                   grid.text("*", x, y, gp = gpar(fontsize = 30, fontface = "bold"))
               }, 
               right_annotation = c(hb1,hb2,hb3))
  results <- c(fct = ht, lgd = lgd)
  return(results)
}

#all Acidobacteriota & Actinobacteriota
j = 1; i = 42
full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,], sig_matrix = sig_mat[j:i,])
png(paste0("../Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_LT_tree_1.png"), width = 2400, height = 300+45*((i-j)))
draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) # add space for titles
draw(full_map$lgd, x = unit(45, "cm"), y = unit(6, "cm"))
dev.off()

#all random
j = 43; i = 82
full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,] ,sig_matrix = sig_mat[j:i,])
png(paste0("../Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_LT_tree_2.png"), width = 2400, height = 300+45*((i-j)))
draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) # add space for titles
draw(full_map$lgd, x = unit(45, "cm"), y = unit(6, "cm"))
dev.off()


#Proteobacteria 1/2
j = 83; i = 136
full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,] ,sig_matrix = sig_mat[j:i,])
png(paste0("../Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_LT_tree_3.png"), width = 2400, height = 300+45*((i-j)))
draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) # add space for titles
draw(full_map$lgd, x = unit(45, "cm"), y = unit(6, "cm"))
dev.off()

#Proteobacteria 2/2
j = 137; i = 189
full_map = heatmap_function(num_matrix = mat.diff[j:i,], omatrix = mat.df[j:i,] ,sig_matrix = sig_mat[j:i,])
png(paste0("../Plots/ANCOM-BC/ASV_heatmap_differentials_all_taxa_LT_tree_4.png"), width = 2400, height = 300+45*((i-j)))
draw(full_map$fct, padding = unit(c(5, 1, 1, 30), "cm")) # add space for titles
draw(full_map$lgd, x = unit(45, "cm"), y = unit(6, "cm"))
dev.off()
```


